---
title: "1) Genotyping Accuracy"
output: html_notebook
---

# Setup 
```{r, message=F}
library(tidyverse)
library(cowplot)
source("data_import_functions.R")
source("figure_format_functions.R")
source("calculation_functions.R")
isb_path <- "/labs/khatrilab/solomonb/covid/isb"
isb_samples <- read_tsv("/labs/khatrilab/solomonb/covid/isb/logs/210217_232725/parallel.log") %>% 
  separate(Command, into = c(NA, "sample"), sep = " ") %>% 
  pull(sample) %>% unique()
```

```{r}
all_hla <- format_hla_table(combine_HLA_import(path = isb_path, samples = isb_samples, expand_invitro = T))
```


# Calculations

### Prediction tally

```{r}
prediction_tally <- all_hla %>% 
  allele_tally() %>% 
  filter(genotyper != "invitro") %>% 
  filter(
    grepl("-BL$", sample),  # Limit to the BL time point, others addressed in another figure 
    !grepl("^DRB[345]", locus))  # DRB345 addressed in another figure
```


### Success
```{r}
success_df <- compare_hla(
  hla_df = all_hla, 
  reference = "invitro", 
  method = "success", 
  penalize_extra_drb345 = F
) %>% 
  filter(
    grepl("-BL$", sample),
    !grepl("^DRB[345]", locus))
```

### Accuracy

```{r,message = F, warning=F}
accuracy_df <- compare_hla(
  hla_df = all_hla, 
  reference = "invitro", 
  method = "accuracy",
  penalize_extra_drb345 = T
) %>% 
  filter(
    grepl("-BL$", sample),
    !grepl("^DRB[345]", locus))
```

# Plotting

### Predicition tally

```{r, fig.width = 5, fig.height = 8}
predictions_plt <- gg_hla_prediction_tally(prediction_tally)
predictions_plt
```
### Success

```{r, fig.width = 5, fig.height = 4}
success_plt <- calculate_summary_df(success_df) %>% 
  gg_summary_hla_accuracy(color_label = "Success")
success_plt
```


### Accuracy

```{r, fig.width = 5, fig.height = 4}
accuracy_plt <-  calculate_summary_df(accuracy_df) %>% 
  gg_summary_hla_accuracy(color_label = "Accuracy")
accuracy_plt
```

# Full figure

```{r,fig.height = 6, fig.width = 9}
plt_heatmaps <- plot_grid(
  success_plt,
  accuracy_plt,
  ncol = 1,
  labels = c("B","C")
)
plot_grid(
  predictions_plt,
  plt_heatmaps,
  labels = c("A",NA)
)
```

# Accuracy table

```{r, fig.width = 12}
calculate_summary_df(accuracy_df) %>% 
  flex_summary_hla_accuracy() %>% 
  fontsize(size = 8, part = "all") %>% 
  print(preview = "pptx")
```

# Success table
```{r}
calculate_summary_df(success_df) %>% 
  flex_summary_hla_accuracy() %>% 
  fontsize(size = 8, part = "all") %>% 
  print(preview = "pptx")
```


# Accuracy by allele

```{r, fig.width = 16, fig.height = 8}
accuracy_df %>% 
gg_allele_hla_accuracy()
```

# Accuracy by individual

```{r, fig.width = 12, fig.height = 5}
accuracy_df %>% 
  gg_individual_hla_accuracy(color = "viridis")
```

```{r, fig.width = 12, fig.height = 5}
sample_accuracy_plot <- readRDS("isb_accuracy_drb345_filtered.RDS") %>% 
  filter(grepl("^[ABC]|^D[PQR][AB]1",locus)) %>% 
  filter(grepl("BL$", sample)) %>% 
  gg_individual_hla_accuracy(color = "viridis")
```

```{r, message = F}
hla_mapping_stats_import_v2 <- function(samples, log_dir){
  tibble(sample = samples) %>% 
    mutate(data = map(sample, function(x){
      log_path <- sprintf("%s/%s.genotype.log",log_dir,x)
      df <- tibble(lines = read_lines(log_path))
      if (any(grepl("error", df$lines, ignore.case = T))){
        NA
      } else {
        total_reads <- df %>% 
          mutate(lines = gsub("\t", "", lines)) %>% 
          filter(grepl("^\\[alignment\\] Processed", lines)) %>%
          mutate(lines = gsub("\\,.*", "", lines)) %>% 
          mutate(lines = str_extract(lines, "[0-9]+")) %>% 
          pull(lines) %>% 
          as.numeric()
        df %>% 
          mutate(lines = gsub("\t", "", lines)) %>% 
          filter(grepl("^HLA", lines)) %>% 
          separate(lines, into = c("locus", "abundance", "reads", "classes"), sep = " +") %>% 
          mutate(total_reads = total_reads) %>% 
          mutate(total_hla_reads = sum(as.numeric(reads)))
      }
    })) %>%
    unnest(data) %>% 
    select(!(contains("data"))) %>% 
    separate(locus, into = c(NA, "locus"), sep = "-") %>% 
    mutate_at(c("reads", "classes"), as.numeric)
}

isb_path <- "/labs/khatrilab/solomonb/covid/isb"
isb_samples <- read_tsv("/labs/khatrilab/solomonb/covid/isb/logs/210217_232725/parallel.log") %>% 
  separate(Command, into = c(NA, "sample"), sep = " ") %>% 
  pull(sample) %>% unique()
arcas_log_dir <- sprintf("%s/arcasHLA", isb_path)
```

### Calculate mapping efficiency 
```{r}
mapping <- hla_mapping_stats_import_v2(isb_samples, arcas_log_dir) %>% 
  filter(grepl("^[ABC]|^D[PQR][AB]1", locus) & grepl("BL$", sample))
```

```{r, fig.width = 12, fig.height = 2}
sample_reads_plot <- mapping %>% 
  select(sample, total_hla_reads) %>% 
  distinct() %>% 
      ggplot(aes(x = sample, y = 1, fill = log(total_hla_reads))) + 
      geom_tile() +
      theme_classic() +
      # facet_grid(locus ~.) +
      scale_fill_viridis_c(na.value = "transparent", option = "viridis") +
      theme(axis.text = element_blank(), axis.ticks = element_blank()) +
      labs(x = "Sample", y = "", fill = "Reads")
```
```{r, fig.width = 12, fig.height = 2}
mapping %>% 
      ggplot(aes(x = sample, y = locus, fill = log(reads))) + 
      geom_tile() +
      theme_classic() +
      # facet_grid(locus ~.) +
      scale_fill_viridis_c(na.value = "transparent", option = "viridis") +
      # scale_y_reverse()+
      theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
      labs(x = "Sample", y = "", fill = "Reads")
```

```{r}
plot_grid(
  sample_reads_plot,
  sample_accuracy_plot,
  ncol = 1,
  align="v",
  axis="lr",
  rel_heights = c(0.2,0.8)
)
```

```{r}
readRDS("isb_accuracy_drb345_filtered.RDS") %>% 
  filter(grepl("^[ABC]|^D[PQR][AB]1",locus)) %>% 
  filter(grepl("BL$", sample)) %>% 
  filter(field == "field_2") %>% 
  group_by(sample) %>% 
  summarise(sum = sum(accuracy)) %>% 
  arrange(sum)
```

```{r}
mapping %>% 
  filter(sample == "INCOV017-BL")
```


# Old work

<!-- ### Updated accuracy with expanded HLA -->

<!-- ```{r, fig.height=8} -->
<!-- ### Every subject has molecular HLA typing, but may occur in the BL, AC, or CV timepoint -->
<!-- ### This expands the molecular HLA typing from the timepoint it occurs to all other time points  -->
<!-- ### This allows for scRNA genotyping from a given subject and timepoint to be assessed even if that -->
<!-- ### subjects molecular typing happened at a different time point. Since HLA is genetic, the molecular -->
<!-- ### typing should not change -->
<!-- original <- format_hla_table(combine_HLA_import(path = isb_path, samples = isb_samples, expand_invitro = F)) -->
<!-- expanded <- format_hla_table(combine_HLA_import(path = isb_path, samples = isb_samples, expand_invitro = T)) -->

<!-- original <- compare_hla(hla_df = original, reference = "invitro", exclude_missing = F, compare_function = "accuracy") -->
<!-- expanded <- compare_hla(hla_df = expanded, reference = "invitro", exclude_missing = F, compare_function = "accuracy") -->


<!-- original %>%  -->
<!--   filter(grepl("^[ABC]", locus)) %>%  -->
<!--   gg_accuracy() + -->
<!--   ggtitle("MHC1 loci accuracy", subtitle = "Original in vitro vs. sample pairing") -->
<!-- expanded %>%  -->
<!--   filter(grepl("BL$", sample)) %>%  -->
<!--   filter(grepl("^[ABC]", locus)) %>%  -->
<!--   gg_accuracy() + -->
<!--   ggtitle("MHC1 loci accuracy", subtitle = "Expanded HLA invitro vs. sample pairings. BL only") -->

<!-- ``` -->
