---
title: "4) Predicting HLA-DRB4 paralogs"
output: html_notebook
---

```{r}
all_hla_expanded <- readRDS("all_hla_expanded.RDS")
```


```{r}
alignment_stats_df <- tibble(sample = isb_samples) %>% 
  mutate(data = map(sample, function(x){
    log_path <- sprintf("%s/logs/210309_155222/%s/%s_arcas_genotype.log",isb_path,x,x)
    df <- tibble(lines = read_lines(log_path))
    if (any(grepl("error", df$lines, ignore.case = T))){
      NA
    } else {
      df %>% 
        mutate(lines = gsub("\t", "", lines)) %>% 
        filter(grepl("^HLA", lines)) %>% 
        separate(lines, into = c("locus", "abundance", "reads", "classes"), sep = " +")
    }
  })) %>%
  unnest(data) %>% 
  select(-data) %>% 
  separate(locus, into = c(NA, "locus"), sep = "-")
```


```{r}
drb_genes_df <- all_hla_expanded %>% 
  filter(grepl("BL$", sample)) %>% # Limit to one time collection, so no repeats
  select(sample, locus, genotyper) %>% 
  filter(grepl("DRB", locus), genotyper %in% c("arcasHLA", "invitro")) %>% 
  distinct() %>% 
  group_by(genotyper, sample) %>% 
  arrange(locus, .by_group = T) %>% 
  summarise(locus = paste(locus, collapse = "-")) %>% 
  pivot_wider(names_from = "genotyper", values_from = "locus", values_fill = NA) %>% 
  drop_na()
```

```{r}
DRB_relative_filter <- function(relative_var="reads", threshold=0.05){
  alignment_stats_df %>% 
  # mutate(abundance = as.numeric(gsub("\\%", "", abundance))) %>% 
  filter(grepl("DRB[1345]", locus)) %>% 
  mutate_at(c('reads', 'classes'), as.numeric) %>% 
  group_by(sample) %>% 
  # mutate(freq = round(sym(relative_var)/sum(sym(relative_var)), digits = 2)) %>% 
  mutate(freq = round(!!sym(relative_var)/sum(!!sym(relative_var)), digits = 2)) %>% 
  select(sample, locus, freq) %>% 
  filter(freq >= threshold) %>% 
  arrange(locus, .by_group = T) %>% 
  summarise(locus = paste(locus, collapse = "-")) 
}

DRB_relative_filter()
```

```{r}
x <- tibble(relative_var = c("reads", "classes")) %>% 
  mutate(data = map(relative_var, function(relative_var){
    tibble(thresh = seq(0,1, by = 0.01)) %>% 
      mutate(data = map(thresh, function(x) DRB_relative_filter(threshold = x, relative_var = relative_var))) %>% 
      unnest(data) %>% 
      pivot_wider(names_from = "thresh", values_from = "locus", names_prefix = "threshold_") %>% 
      right_join(drb_genes_df, by = "sample") %>% 
      mutate_at(vars(contains(c("threshold", "arcas"))), function(x) x == .$invitro) %>% 
      ungroup() %>% 
      summarise_if(is.logical, function(x) sum(x, na.rm = T)/length(x))
  })) %>% 
  unnest(data) %>% 
  pivot_longer(!relative_var, names_to = "threshold", values_to = "accuracy", names_prefix = "threshold_") %>% 
  {.$accuracy[.$threshold=="arcasHLA"][1]->>arcas;.} %>% 
  filter(threshold != "arcasHLA") %>% 
  mutate(threshold = as.numeric(threshold)) 


x %>% 
  ggplot(aes(x = threshold, y = accuracy, color = relative_var))+
  geom_path()+
  geom_hline(yintercept = arcas, color = "red")+
  geom_point()+
  coord_cartesian(ylim = c(0, 1))+
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90)) +
  scale_y_continuous(n.breaks = 10)+
  scale_color_brewer(palette = "Dark2")
# +
  # ggtitle("Accuracy of genes in DRB loci", subtitle = sprintf("Gene inclusion genotype based on relative %s threshold", relative_var))
```

