---
title: "4) Predicting HLA-DRB4 paralogs"
output: html_notebook
---

# Startup

```{r}
library(tidyverse)
source("data_import_functions.R")
all_hla_expanded <- readRDS("all_hla_expanded.RDS")
isb_path <- "/labs/khatrilab/solomonb/covid/isb"
isb_samples <- read_tsv("/labs/khatrilab/solomonb/covid/isb/logs/210217_232725/parallel.log") %>% 
  separate(Command, into = c(NA, "sample"), sep = " ") %>% 
  pull(sample) %>% unique()
```


```{r}
arcas_log_dir <- sprintf("%s/logs/210309_155222",isb_path)
alignment_stats_df <- hla_mapping_stats_import(isb_samples, arcas_log_dir)
```

# Determining DRB inclusion threshold

```{r}
# Generate df with set of DRB genes determined for each sample from 
# arcasHLA alignment vs. molecular typing
drb_genes_df <- all_hla_expanded %>% 
  filter(grepl("BL$", sample)) %>% # Limit to one time collection, so no repeats
  select(sample, locus, genotyper) %>% 
  filter(grepl("DRB", locus), genotyper %in% c("arcasHLA", "invitro")) %>% 
  distinct() %>% 
  group_by(genotyper, sample) %>% 
  arrange(locus, .by_group = T) %>% 
  summarise(locus = paste(locus, collapse = "-")) %>% 
  pivot_wider(names_from = "genotyper", values_from = "locus", values_fill = NA) %>% 
  drop_na()
drb_genes_df
```

```{r}
# Generates set of DRB genes with number of mapped reads reaching a 
# specified threshold. Depends on alignment_stats_df
DRB_relative_filter <- function(relative_var="reads", threshold=0.05){
  alignment_stats_df %>% 
  filter(grepl("DRB[1345]", locus)) %>% 
  mutate_at(c('reads', 'classes'), as.numeric) %>% 
  group_by(sample) %>% 
  mutate(freq = round(!!sym(relative_var)/sum(!!sym(relative_var)), digits = 2)) %>% 
  select(sample, locus, freq) %>% 
  filter(freq >= threshold) %>% 
  arrange(locus, .by_group = T) %>% 
  summarise(locus = paste(locus, collapse = "-")) 
}
DRB_relative_filter()
```

```{r}
DRB_threshold_accuracy <- tibble(relative_var = c("reads", "classes")) %>% 
  mutate(data = map(relative_var, function(relative_var){
    tibble(thresh = seq(0,1, by = 0.01)) %>% 
      mutate(data = map(thresh, function(x) DRB_relative_filter(threshold = x, relative_var = relative_var))) %>% 
      unnest(data) %>% 
      pivot_wider(names_from = "thresh", values_from = "locus", names_prefix = "threshold_") %>% 
      right_join(drb_genes_df, by = "sample") %>% 
      mutate_at(vars(contains(c("threshold", "arcas"))), function(x) x == .$invitro) %>% 
      ungroup() %>% 
      summarise_if(is.logical, function(x) sum(x, na.rm = T)/length(x))
  })) %>% 
  unnest(data) %>% 
  pivot_longer(!relative_var, names_to = "threshold", values_to = "accuracy", names_prefix = "threshold_") %>% 
  # Output accuracy with no threshold
  {.$accuracy[.$threshold=="arcasHLA"][1]->>baseline_arcas_accuracy;.} %>% 
  filter(threshold != "arcasHLA") %>% 
  mutate(threshold = as.numeric(threshold)) 
DRB_threshold_accuracy
```

```{r}
DRB_threshold_accuracy  %>% 
  ggplot(aes(x = threshold, y = accuracy, color = relative_var))+
  geom_path()+
  geom_hline(yintercept = baseline_arcas_accuracy, color = "red")+
  geom_point()+
  coord_cartesian(ylim = c(0, 1))+
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90)) +
  scale_y_continuous(n.breaks = 10)+
  scale_color_brewer(palette = "Dark2")
```

# Comparing effect on overall DRB accuracy

```{r}
# Create a version of all_hla_expanded for only DRB3/4/5 where there is an entry
# for DRB3, 4, and 5 for invitro, arcasHLA, and hlaminer. If there is no prediction
# for a resulting entry, "X" is input so that if invitro and a genotyper have an "X"
# that would be considered a match.

drb_testing <- expand.grid(sample = all_hla_expanded %>% pull(sample) %>% unique(), 
            allele_id = c("1","2"),
            locus = c("DRB3", "DRB4", "DRB5"),
            genotyper = c("invitro", "arcasHLA", "hlaminer")) %>% 
  arrange(sample, locus, genotyper, allele_id) %>% 
  left_join(all_hla_expanded %>% filter(genotyper %in% c("invitro", "arcasHLA", "hlaminer")), join_by = c("sample", "allele_id", "locus", "genotyper"))
  # # Fill in "X"s for NA resulting from the expansion, but not because of original
  # # failed genotyping predictions
  # mutate(field_1 = ifelse(is.na(field_1), "X", field_1)) %>% 
  # mutate_at(vars(field_2:field_3), funs(ifelse(is.na(.) & field_1 == "X", "X", .)))
drb_testing
```

```{r, warning = F}
compare_hla(hla_df = drb_testing %>% filter(grepl("BL$", sample)), 
            reference = "invitro", 
            exclude_missing = F, 
            compare_function = "accuracy") %>% 
  gg_accuracy()
```


```{r}
DRB_gene_predictions <- DRB_relative_filter(relative_var="reads", threshold=0.05)
DRB_gene_predictions
```

```{r, warning=F}
x <- compare_hla(hla_df = drb_testing %>% filter(grepl("BL$", sample)), 
            reference = "invitro", 
            exclude_missing = F, 
            compare_function = "accuracy",
            na_drop = F) %>% 
  left_join(valid_DRB, join_by = c("sample", "locus")) %>% 
  mutate(valid_invitro = map_lgl(reference, function(x) any(!grepl("X", x)))) %>% 
  filter(valid_DRB == T | valid_invitro == T)
```

```{r}
any(!grepl("X", c("01","X")))
```


```{r}
valid_DRB <- DRB_gene_predictions %>% mutate(locus = str_split(locus, "-")) %>% unnest(locus) %>% mutate(valid_DRB = T)
x <- hla_compare_composite %>% left_join(valid_DRB, join_by = c("sample", "locus")) %>% mutate(not_DRB = !grepl("DRB", locus))
```

