---
title: "4) Predicting HLA-DRB4 paralogs"
output: html_notebook
---

# Startup

```{r, message = F}
library(tidyverse)
library(tidymodels)
library(cowplot)
library(kknn)
source("data_import_functions.R")
source("calculation_functions.R")
source("ggplot_layouts.R")
all_hla_expanded <- readRDS("all_hla_expanded.RDS")
isb_path <- "/labs/khatrilab/solomonb/covid/isb"
isb_samples <- read_tsv("/labs/khatrilab/solomonb/covid/isb/logs/210217_232725/parallel.log") %>% 
  separate(Command, into = c(NA, "sample"), sep = " ") %>% 
  pull(sample) %>% unique()
```


```{r}
arcas_log_dir <- sprintf("%s/logs/210309_155222",isb_path)
alignment_stats_df <- hla_mapping_stats_import(isb_samples, arcas_log_dir)
```


# DRB gene copy number by KNN

- The above strategy only determines if a locus should be included/excludes, but omits two additional considerations:
    - The number of alleles of a given DRB gene can vary from 0 to 2 copies
    - The total number of DRB3 + 4 + 5 alleles has a maximum of two
- This allows several non-biological outcomes
    - Possible to have three loci meet threshold, which would result in a minimum of 3 total HLA-DRB345 alleles
    - Even if only 2 loci meet threshold, if two alleles are predicted for each, total number of alleles would be 4
- Can we predict the copy number for each gene based on read counts?
    - PMID 28145101 reported success with KNN applied to ratio of DRB345:DRB1 reads
    

### Demonstrating prediction frequencies

```{r, fig.width = 6, fig.height=5}
all_hla <- readRDS("all_hla_expanded.RDS")

pre_tally_plt <- all_hla %>%   
  allele_tally() %>% 
  filter(grepl("DRB[1345]", locus), genotyper %in% c("invitro", "arcasHLA", "hlaminer")) %>% 
  gg_hla_prediction_tally()
pre_tally_plt
```




### Pre-processing and visualization

```{r}
# Calculate ratio of DRB345/DRB1 reads for each sample
drb345_ratios <- alignment_stats_df %>% filter(grepl("^DRB[1345]", locus)) %>% 
  select(sample, locus, reads) %>% 
  mutate(reads = as.numeric(reads)) %>% 
  pivot_wider(names_from = "locus", values_from = 'reads', values_fill = 0) %>% 
  mutate_at(vars(DRB3:DRB5), funs(./DRB1)) %>% 
  select(-DRB1) %>% 
  pivot_longer(!sample, names_to = 'locus', values_to ='frequency')
drb345_ratios
```

```{r}
# Determine ground truth DRB345 copy number
invitro <- invitro_import()
drb345_copy_number <- invitro %>% separate(allele, into=c('locus','allele'), sep = '\\*') %>% filter(grepl('^DRB[345]',locus)) %>%
  select(-allele, -genotyper) %>% 
  pivot_wider(names_from ='locus', values_from = 'allele_id', values_fn = length, values_fill = 0) %>% 
  pivot_longer(!sample, names_to = 'locus', values_to = 'copy_number') %>% 
  separate(sample, into = c('sample',NA), sep = '-') 
drb345_copy_number
```

```{r}
# Join DRB345 ratios with molecular typing copy numbers
drb345_df <- drb345_ratios %>% 
  separate(sample, into = c('sample','time'), sep = '-') %>% 
  pivot_wider(names_from = 'locus', values_from ='frequency', values_fill=0) %>% 
  left_join(drb345_copy_number, by = c('sample')) %>% 
  filter(!grepl("CV$",time)) %>% # Exclude potential CV timepoint weirdness
  drop_na() # Drop the few samples with molecular typing but no scRNA
drb345_df
```

```{r}
# Frequency of DRB345:DRB1 ratio relative to ground truth copy number
drb345_df %>% 
  pivot_longer(DRB3:DRB5, names_to='locus_2', values_to='drb345_ratio') %>% 
  filter(locus == locus_2) %>% 
  ggplot(aes(x=drb345_ratio, color = locus, fill = locus))+
  geom_density(alpha = 0.1) +
  theme_bw() +
  scale_fill_brewer(palette = "Set1") + scale_color_brewer(palette = "Set1") +
  facet_grid(copy_number ~ locus, scales = "free")
```
### KNN model

```{r}
# Train/test split
set.seed(123)
drb_split <- initial_split(drb345_df %>% mutate(copy_number = factor(copy_number)), prop = 0.7, strata = copy_number)
drb_train <- drb_split %>% training()
drb_test <- drb_split %>% testing()
```

```{r}
# Model design and fit
drb_model <- nearest_neighbor() %>%
  set_engine("kknn") %>%
  set_mode("classification")

drb_fit <- drb_model %>% fit(copy_number ~locus+DRB3+DRB4+DRB5, data = drb_train)

## Alternative models
# drb_model <- rand_forest(trees = 1000) %>%
#   set_mode("classification") %>%
#   set_engine("ranger")
# drb_model <- discrim::naive_Bayes() %>% 
#   set_engine("naivebayes")
```

### Performance on training data
```{r}
# Model predictions
drb_train_outcome <- drb_train %>% 
  bind_cols(
    drb_fit %>% predict(new_data = drb_train), # Adds predicted class
    drb_fit %>% predict(new_data = drb_train, type = 'prob') # Adds class predicition probabilities
  )

drb_train_outcome

# ROC plots
training_plt <- gg_multilevel_roc(drb_train_outcome)
training_plt

# AUC calculation
drb_train_outcome %>% roc_auc(truth = copy_number, .pred_0:.pred_2)
```

### Performance on testing data

```{r}
# Model predictions
drb_test_outcome <- drb_test %>% 
  bind_cols(
    drb_fit %>% predict(new_data = drb_test), 
    drb_fit %>% predict(new_data = drb_test, type = 'prob') 
  )

drb_test_outcome

# ROC plots
testing_plt <- gg_multilevel_roc(drb_test_outcome)
testing_plt

# AUC calculation
drb_test_outcome %>% roc_auc(truth = copy_number, .pred_0:.pred_2)
```


# Update DBR genotypes

```{r}
# Predictions from full data
drb_key <- drb345_df %>% bind_cols(
  drb_fit %>% predict(new_data = drb345_df)
) %>% 
  select(sample, time, locus, copy_number, .pred_class) %>% 
  unite(sample, sample, time, sep = "-")
drb_key
```
### Predicted allele tally

```{r, fig.width = 3, fig.height=5}
post_tally_plt <- drb_key %>% 
  rename("copy_invitro" = copy_number, "copy_predicted" = .pred_class) %>% 
  pivot_longer(contains("copy_"), names_to = "genotyper", values_to = "count", values_transform = list(count = function(x) as.numeric(as.character(x))), names_prefix = "copy_") %>% 
  mutate(field = " ") %>% 
  gg_hla_prediction_tally()
post_tally_plt
```

#### Filtered DRB accuracy

```{r}
# Select best arcasHLA allele genotype up to maximum number predicted by knn model 
## allele_id is in order of best match per locus, so filtering to keep only
## those allele_ids lower or equal to the .pred_class will keep only the
## best alleles up to the allowable number for a given locus 
drb345_filtered <- all_hla %>% 
  filter(genotyper == "arcasHLA" & grepl("^DRB[345]", locus)) %>% 
  left_join(drb_key, by = c("sample", "locus")) %>% 
  mutate(.pred_class = as.character(.pred_class)) %>% 
  mutate_at(vars(c("copy_number", ".pred_class")), function(x) ifelse(is.na(x), "0", x)) %>% 
  filter(allele_id <= .pred_class) 
drb345_filtered
```

```{r}
# Create a version of all_hla_expanded for only DRB3/4/5 where there is an entry
# for DRB3, 4, and 5 for invitro, arcasHLA, and hlaminer. If there is no prediction
# for a resulting entry, "X" is input so that if invitro and a genotyper have an "X"
# that would be considered a match.

drb_testing <- 
  # Create grid of all samples, allele_id, locus, and genotyper combinations
  expand.grid(sample = all_hla %>% pull(sample) %>% unique(), 
                           allele_id = c("1","2"),
                           locus = c("DRB1", "DRB3", "DRB4", "DRB5"),
                           genotyper = c("invitro", "unfiltered", "filtered")) %>% 
  arrange(sample, locus, genotyper, allele_id) %>%
  # Add in invitro, original arcasHLA, and filtered arcasHLA predictions by joining
  left_join(
    bind_rows(
      all_hla %>% filter(genotyper == "invitro" & grepl("^DRB[1345]", locus)),
      all_hla %>% filter(genotyper == "arcasHLA" & grepl("^DRB[1345]", locus)) %>% mutate(genotyper = "unfiltered"), 
      drb345_filtered %>% select(-copy_number, -.pred_class) %>% mutate(genotyper = "filtered")), 
    join_by = c("sample", "allele_id", "locus", "genotyper")) %>% 
  # Fill in "X"s for NA resulting from the expansion, but not because of original
  # failed genotyping predictions
  filter(!is.na(field_1))
  # mutate(field_1 = ifelse(is.na(field_1), "X", field_1)) %>%
  # mutate_at(vars(field_2:field_3), funs(ifelse(is.na(.) & field_1 == "X", "X", .)))
drb_testing
```

```{r}
drb345_accuracy <- compare_hla(hla_df = drb_testing %>% filter(grepl("BL$", sample)),
            reference = "invitro",
            exclude_missing = F,
            penalize_extra_drb345 = T) %>%
  mutate(genotyper = fct_relevel(genotyper, "unfiltered"))
drb345_accuracy
```


```{r, fig.width = 5, fig.height = 3.5}
accuracy_plt <- calculate_summary_df(drb345_accuracy) %>% 
  gg_summary_hla_accuracy(color_label = "Accuracy")
accuracy_plt
```


```{r}
# saveRDS(drb345_accuracy, "drb345_accuracy.RDS")
```

# Combined plot

```{r, fig.width = 12, fig.height = 4}
combined_plt <- cowplot::plot_grid(
  training_plt,
  testing_plt,
  accuracy_plt,
  nrow = 1,
  align = "h", axis = "bt",
  rel_widths = c(4,4,4)
)
cowplot::plot_grid(
  cowplot::plot_grid(
    NA,NA,NA,
    labels = c("Training", "Testing", "Accuracy"),
    nrow = 1
  ),
  combined_plt,
  ncol = 1,
  rel_heights = c(0.07, 0.95)
)
```

```{r, fig.width = 7, fig.height = 6}

plot_grid(
  plot_grid(
    pre_tally_plt, testing_plt,
    nrow = 1,
    align = "h",
    axis = "bt",
    labels = c("A","B"),
    rel_widths = c(4,3)
  ),
  plot_grid(
    post_tally_plt,
    accuracy_plt,
    nrow = 1,
    align = "h",
    axis = "bt",
    rel_widths = c(1,2),
    labels = c("C","D")
  ),
  ncol = 1
)

```

# Output updated all_hla data

```{r}
drb345_filtered <- all_hla %>% 
  filter(genotyper != "invitro" & grepl("^DRB[345]", locus)) %>% 
  left_join(drb_key, by = c("sample", "locus")) %>% 
  mutate(.pred_class = as.character(.pred_class)) %>% 
  mutate_at(vars(c("copy_number", ".pred_class")), function(x) ifelse(is.na(x), "0", x)) %>% 
  filter(allele_id <= .pred_class) 
drb345_filtered

unique(drb345_filtered$genotyper)

drb_testing <- 
  # Create grid of all samples, allele_id, locus, and genotyper combinations
  expand.grid(sample = all_hla %>% pull(sample) %>% unique(), 
                           allele_id = c("1","2"),
                           locus = c("DRB1", "DRB3", "DRB4", "DRB5"),
                           genotyper = c("invitro", unique(drb345_filtered$genotyper)),
                           filtered = c(T,F)) %>% 
  arrange(sample, locus, genotyper, allele_id) %>%
  # Add in invitro, original arcasHLA, and filtered arcasHLA predictions by joining
  left_join(
    bind_rows(
      all_hla %>% filter(genotyper == "invitro" & grepl("^DRB[1345]", locus)),
      all_hla %>% filter(grepl("^DRB[1345]", locus)) %>% mutate(filtered = F),
      drb345_filtered %>% select(-copy_number, -.pred_class) %>% mutate(filtered = T)), 
    join_by = c("sample", "allele_id", "locus", "genotyper", "filtered")) %>% 
  # Fill in "X"s for NA resulting from the expansion, but not because of original
  # failed genotyping predictions
    filter(!is.na(field_1))
drb_testing

drb345_accuracy <- compare_hla(hla_df = drb_testing %>% filter(grepl("BL$", sample)) %>% unite(genotyper, filtered, col = "genotyper", sep = "_"),
            reference = "invitro_FALSE",
            exclude_missing = F) %>%
  mutate(genotyper = fct_relevel(genotyper, "unfiltered"))
drb345_accuracy

calculate_summary_df(drb345_accuracy) %>% 
  separate(genotyper, into = c("genotyper2", "genotyper"), sep = "_") %>% 
  mutate(genotyper = fct_recode(genotyper, "Unfiltered" = "FALSE", "Filtered" = "TRUE")) %>% 
  mutate(genotyper2 = reformat_hla_genotyper(genotyper2)) %>% 
  gg_summary_hla_accuracy(color_label = "Accuracy") + facet_grid(genotyper2~field) + theme(strip.text.y = element_text())

```
```{r,message = F, warning=F}
accuracy_df <- compare_hla(hla_df = all_hla, 
            reference = "invitro", 
            exclude_missing = F, 
            compare_function = "accuracy") 
calculate_summary_df(accuracy_df) %>% 
  gg_summary_hla_accuracy(color_label = "Accuracy")
```
```{r}
x <- accuracy_df %>% filter(field == "field_2", genotyper == "hlaminer", locus == "DRB4")
```
```{r}
y <- drb345_accuracy %>% filter(field == "field_2", genotyper == "hlaminer_FALSE", locus == "DRB4")
```

# Validation on PMID

- Ultimately, performance is lower because there are fewer DRB345 reads overall 
- Since this reflects a different sequencing modality, it does not seem to be a fair validation

<!-- ```{r} -->
<!-- pmid_log_dir <- "/labs/khatrilab/solomonb/covid/pmid30518681/scRNA/arcasHLA" -->
<!-- pmid_samples <- read_tsv("/labs/khatrilab/solomonb/covid/pmid30518681/scRNA/SRR_Acc_List.txt", col_names = "srr") %>% pull(srr) -->
<!-- pmid_alignment_stats_df <- tibble(sample = pmid_samples) %>%  -->
<!--     mutate(data = map(sample, function(x){ -->
<!--       log_path <- sprintf("%s/%s/%s_arcas_genotype.log",pmid_log_dir,x,x) -->
<!--       df <- tibble(lines = read_lines(log_path)) -->
<!--       if (any(grepl("error", df$lines, ignore.case = T))){ -->
<!--         NA -->
<!--       } else { -->
<!--         df %>%  -->
<!--           mutate(lines = gsub("\t", "", lines)) %>%  -->
<!--           filter(grepl("^HLA", lines)) %>%  -->
<!--           separate(lines, into = c("locus", "abundance", "reads", "classes"), sep = " +") -->
<!--       } -->
<!--     })) %>% -->
<!--     unnest(data) %>%  -->
<!--     # select(-data) %>%  -->
<!--     separate(locus, into = c(NA, "locus"), sep = "-") -->
<!-- pmid_alignment_stats_df -->

<!-- pmid_drb345_ratios <- pmid_alignment_stats_df %>% filter(grepl("^DRB[1345]", locus)) %>%  -->
<!--   select(sample, locus, reads) %>%  -->
<!--   mutate(reads = as.numeric(reads)) %>%  -->
<!--   pivot_wider(names_from = "locus", values_from = 'reads', values_fill = 0) %>%  -->
<!--   mutate_at(vars(DRB3:DRB5), funs(./DRB1)) %>%  -->
<!--   select(-DRB1) %>%  -->
<!--   pivot_longer(!sample, names_to = 'locus', values_to ='frequency') -->
<!-- pmid_drb345_ratios -->

<!-- pmid_invitro <- invitro_hla <- invitro_import(path = "pmid30518681_scisco_hla.tsv", exclude_comment_samples = T) %>%  -->
<!--   mutate(sample = gsub("STU-HD-","",sample)) -->
<!-- pmid_invitro -->

<!-- pmid_drb345_copy_number <- pmid_invitro %>%  -->
<!--   separate(allele, into=c('locus','allele'), sep = '\\*') %>%  -->
<!--   filter(grepl('^DRB[345]',locus)) %>% -->
<!--   select(-allele, -genotyper) %>%  -->
<!--   pivot_wider(names_from ='locus', values_from = 'allele_id', values_fn = length, values_fill = 0) %>%  -->
<!--   pivot_longer(!sample, names_to = 'locus', values_to = 'copy_number') %>%  -->
<!--   separate(sample, into = c('sample',NA), sep = '-')  -->
<!-- pmid_drb345_copy_number -->

<!-- pmid_drb345_df <- pmid_drb345_ratios %>%  -->
<!--   pivot_wider(names_from = 'locus', values_from ='frequency', values_fill=0) %>% -->
<!--   rename(Run = sample) %>%  -->
<!--   left_join(read_csv("/labs/khatrilab/solomonb/covid/pmid30518681/pmid30518681_samples.csv") %>% select(Run, sample), by = "Run") %>%  -->
<!--   mutate(sample = gsub("[0-9]","",sample)) %>%  -->
<!--   left_join(pmid_drb345_copy_number, by = c('sample')) %>%  -->
<!--   drop_na() # Drop the few samples with molecular typing but no scRNA -->
<!-- pmid_drb345_df -->

<!-- pmid_drb_outcome <- pmid_drb345_df %>%  -->
<!--   bind_cols( -->
<!--     drb_fit %>% predict(new_data = pmid_drb345_df),  -->
<!--     drb_fit %>% predict(new_data = pmid_drb345_df, type = 'prob')  -->
<!--   ) -->

<!-- pmid_drb_outcome -->

<!-- roc_plot <- pmid_drb_outcome %>%  -->
<!--   roc_curve(truth = factor(copy_number, levels = c("0","1","2")), .pred_0:.pred_2) %>%  -->
<!--   autoplot() -->
<!-- ggplot_build(roc_plot)$data[[1]] %>%  -->
<!--   mutate(n_alleles = fct_recode(PANEL, "0"="1", "1"="2", "2"="3")) %>%  -->
<!--   ggplot(aes(x=x,y=y,color=n_alleles))+ -->
<!--   geom_path()+ -->
<!--   geom_abline(slop=1,intercept=0, linetype="dotted")+ -->
<!--   theme_bw() + -->
<!--   scale_color_brewer(palette = "Set1") + -->
<!--   labs(x = "1 - Specificity", y = "Sensitivity") -->
<!-- pmid_drb_outcome %>% roc_auc(truth = factor(copy_number, levels = c("0","1","2")), .pred_0:.pred_2) -->
<!-- ``` -->


<!-- ```{bash} -->
<!-- <!-- for i in $(cat /labs/khatrilab/solomonb/covid/pmid30518681/scRNA/SRR_Acc_List.txt) --> -->
<!-- <!-- do --> -->
<!-- <!--   mkdir /labs/khatrilab/solomonb/covid/pmid30518681/scRNA/arcasHLA/${i} --> -->
<!-- <!--   cp /labs/khatrilab/solomonb/covid/pmid30518681/scRNA/arcasHLA/${i}.genotype.log /labs/khatrilab/solomonb/covid/pmid30518681/scRNA/arcasHLA/${i}/${i}_arcas_genotype.log  --> -->
<!-- <!-- done --> -->
<!-- ``` -->


# Old work

<!-- ### Error bar plots -->
<!-- ```{r, warning = F, fig.width=8, fig.height=4} -->
<!-- drb1_plot <- drb345_accuracy %>% filter(locus == "DRB1" & genotyper == "unfiltered") %>% gg_accuracy() -->
<!-- drb345_plot <- drb345_accuracy %>% filter(locus != "DRB1") %>% gg_accuracy() -->
<!-- drb1_plot | drb345_plot | plot_layout(widths = c(3,10)) -->
<!-- ``` -->

<!-- ### DRB gene inclusion by threshold -->

<!-- ```{r} -->
<!-- # Generate df with set of DRB genes determined for each sample from  -->
<!-- # arcasHLA alignment vs. molecular typing -->
<!-- drb_genes_df <- all_hla_expanded %>%  -->
<!--   filter(grepl("BL$", sample)) %>% # Limit to one time collection, so no repeats -->
<!--   select(sample, locus, genotyper) %>%  -->
<!--   filter(grepl("DRB", locus), genotyper %in% c("arcasHLA", "invitro")) %>%  -->
<!--   distinct() %>%  -->
<!--   group_by(genotyper, sample) %>%  -->
<!--   arrange(locus, .by_group = T) %>%  -->
<!--   summarise(locus = paste(locus, collapse = "-")) %>%  -->
<!--   pivot_wider(names_from = "genotyper", values_from = "locus", values_fill = NA) %>%  -->
<!--   drop_na() -->
<!-- drb_genes_df -->
<!-- ``` -->

<!-- ```{r} -->
<!-- # Generates set of DRB genes with number of mapped reads reaching a  -->
<!-- # specified threshold. Depends on alignment_stats_df -->
<!-- DRB_relative_filter <- function(relative_var="reads", threshold=0.05){ -->
<!--   alignment_stats_df %>%  -->
<!--   filter(grepl("DRB[1345]", locus)) %>%  -->
<!--   mutate_at(c('reads', 'classes'), as.numeric) %>%  -->
<!--   group_by(sample) %>%  -->
<!--   mutate(freq = round(!!sym(relative_var)/sum(!!sym(relative_var)), digits = 2)) %>%  -->
<!--   select(sample, locus, freq) %>%  -->
<!--   filter(freq >= threshold) %>%  -->
<!--   arrange(locus, .by_group = T) %>%  -->
<!--   summarise(locus = paste(locus, collapse = "-"))  -->
<!-- } -->
<!-- DRB_relative_filter() -->
<!-- ``` -->

<!-- ```{r} -->
<!-- DRB_threshold_accuracy <- tibble(relative_var = c("reads", "classes")) %>%  -->
<!--   mutate(data = map(relative_var, function(relative_var){ -->
<!--     tibble(thresh = seq(0,1, by = 0.01)) %>%  -->
<!--       mutate(data = map(thresh, function(x) DRB_relative_filter(threshold = x, relative_var = relative_var))) %>%  -->
<!--       unnest(data) %>%  -->
<!--       pivot_wider(names_from = "thresh", values_from = "locus", names_prefix = "threshold_") %>%  -->
<!--       right_join(drb_genes_df, by = "sample") %>%  -->
<!--       mutate_at(vars(contains(c("threshold", "arcas"))), function(x) x == .$invitro) %>%  -->
<!--       ungroup() %>%  -->
<!--       summarise_if(is.logical, function(x) sum(x, na.rm = T)/length(x)) -->
<!--   })) %>%  -->
<!--   unnest(data) %>%  -->
<!--   pivot_longer(!relative_var, names_to = "threshold", values_to = "accuracy", names_prefix = "threshold_") %>%  -->
<!--   # Output accuracy with no threshold -->
<!--   {.$accuracy[.$threshold=="arcasHLA"][1]->>baseline_arcas_accuracy;.} %>%  -->
<!--   filter(threshold != "arcasHLA") %>%  -->
<!--   mutate(threshold = as.numeric(threshold))  -->
<!-- DRB_threshold_accuracy -->
<!-- ``` -->

<!-- ```{r} -->
<!-- DRB_threshold_accuracy  %>%  -->
<!--   ggplot(aes(x = threshold, y = accuracy, color = relative_var))+ -->
<!--   geom_path()+ -->
<!--   geom_hline(yintercept = baseline_arcas_accuracy, color = "red")+ -->
<!--   geom_point()+ -->
<!--   coord_cartesian(ylim = c(0, 1))+ -->
<!--   theme_bw() + -->
<!--   theme(axis.text.x = element_text(angle = 90)) + -->
<!--   scale_y_continuous(n.breaks = 10)+ -->
<!--   scale_color_brewer(palette = "Dark2") -->
<!-- ``` -->
















<!-- ```{r} -->
<!-- DRB_gene_predictions <- DRB_relative_filter(relative_var="reads", threshold=0.05) -->
<!-- DRB_gene_predictions -->
<!-- valid_DRB <- DRB_gene_predictions %>% mutate(locus = str_split(locus, "-")) %>% unnest(locus) %>% mutate(valid_DRB = T) -->

<!-- ``` -->

<!-- ```{r, warning=F} -->
<!-- x <- compare_hla(hla_df = drb_testing %>% filter(grepl("BL$", sample)),  -->
<!--             reference = "invitro",  -->
<!--             exclude_missing = F,  -->
<!--             compare_function = "accuracy", -->
<!--             na_drop = F) %>%  -->
<!--   left_join(valid_DRB, join_by = c("sample", "locus")) %>%  -->
<!--   mutate(valid_invitro = map_lgl(reference, function(x) any(!is.na(x))))  -->

<!-- %>%  -->
<!--   filter(valid_DRB == T | valid_invitro == T) -->
<!-- x -->
<!-- ``` -->




<!-- ```{r} -->
<!-- y <- alignment_stats_df %>% filter(grepl("^DRB[1345]", locus)) %>%  -->
<!--   select(sample, locus, reads) %>%  -->
<!--   mutate(reads = as.numeric(reads)) %>%  -->
<!--   pivot_wider(names_from = "locus", values_from = 'reads', values_fill = 0) %>%  -->
<!--   mutate_at(vars(DRB3:DRB5), funs(./DRB1)) %>%  -->
<!--   select(-DRB1) %>%  -->
<!--   pivot_longer(!sample, names_to = 'locus', values_to ='frequency') -->
<!-- y -->
<!-- ``` -->
<!-- ```{r} -->
<!-- y %>%  -->
<!--   ggplot(aes(x=frequency, color = locus))+ -->
<!--   geom_density() + -->
<!--   theme_bw() -->
<!-- ``` -->

<!-- ```{r} -->
<!-- invitro <- invitro_import() -->
<!-- drb345_copy_number <- invitro %>% separate(allele, into=c('locus','allele'), sep = '\\*') %>% filter(grepl('^DRB[345]',locus)) %>% -->
<!--   select(-allele, -genotyper) %>%  -->
<!--   pivot_wider(names_from ='locus', values_from = 'allele_id', values_fn = length, values_fill = 0) %>%  -->
<!--   pivot_longer(!sample, names_to = 'locus', values_to = 'copy_number') %>%  -->
<!--   separate(sample, into = c('sample',NA), sep = '-')  -->
<!-- drb345_copy_number -->
<!-- ``` -->

<!-- ```{r} -->
<!-- z <- y %>%  -->
<!--   separate(sample, into = c('sample','time'), sep = '-') %>%  -->
<!--   pivot_wider(names_from = 'locus', values_from ='frequency', values_fill=0) %>%  -->
<!--   left_join(drb345_copy_number, by = c('sample')) -->
<!-- z <- z %>% filter(!grepl("CV$",time)) %>% drop_na()  -->
<!-- z -->

<!--   # ggplot(aes(x = factor(copy_number), y = frequency))+ -->
<!--   # geom_violin() -->
<!--   # ggplot(aes(x=frequency,color=factor(copy_number)))+ -->
<!--   # geom_density(adjust=1)+ -->
<!--   # # facet_grid(.~locus)+ -->
<!--   # theme_bw() -->
<!-- ``` -->

<!-- ```{r} -->
<!-- # z %>% filter(!grepl("CV$",time)) %>% drop_na() %>%  -->
<!-- #   ggplot(aes(x=frequency))+ -->
<!-- #   geom_density()+ -->
<!-- #   facet_wrap(~copy_number, scales="free_y") -->
<!-- ``` -->

<!-- ```{r} -->
<!-- drb_split <- initial_split(z %>% mutate(copy_number = factor(copy_number)), prop = 0.7, strata = copy_number) -->
<!-- drb_train <- drb_split %>% training() -->
<!-- drb_test <- drb_split %>% testing() -->
<!-- ``` -->
<!-- ```{r} -->
<!-- drb_model <- nearest_neighbor() %>% -->
<!--   set_engine("kknn") %>% -->
<!--   set_mode("classification") -->
<!-- # drb_model <- rand_forest(trees = 1000) %>% -->
<!-- #   set_mode("classification") %>% -->
<!-- #   set_engine("ranger") -->
<!-- # drb_model <- discrim::naive_Bayes() %>%  -->
<!-- #   set_engine("naivebayes") -->
<!-- drb_fit <- drb_model %>%  -->
<!--   fit(copy_number ~locus+DRB3+DRB4+DRB5, data = drb_train) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- drb_predict <- predict(drb_fit, new_data = drb_test) -->
<!-- drb_prob <- predict(drb_fit, new_data = drb_test, type='prob') -->
<!-- drb_outcome <- drb_test %>% bind_cols(drb_predict,drb_prob) -->
<!-- drb_outcome -->
<!-- ``` -->
<!-- ```{r} -->
<!-- conf_mat(drb_outcome, -->
<!--          truth = copy_number, -->
<!--          estimate = .pred_class) %>%  -->
<!--   autoplot(type = 'mosaic') -->
<!-- ``` -->
<!-- ```{r} -->
<!-- drb_outcome %>%  -->
<!--   roc_curve(truth = copy_number, .pred_0:.pred_2) %>%  -->
<!--   autoplot() -->
<!-- ``` -->
<!-- ```{r} -->
<!-- drb_outcome %>%  -->
<!--   roc_auc(truth = copy_number, .pred_0:.pred_2) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- drb_outcome -->
<!-- ``` -->
<!-- ```{r} -->
<!-- arcas_copy_df <- all_hla_expanded %>% filter(genotyper == "arcasHLA" & grepl("DRB[345]", locus)) %>%  -->
<!--   ungroup() %>%  -->
<!--   count(sample,locus, name = "arcas_copy_number") %>%  -->
<!--   mutate(arcas_copy_number = factor(arcas_copy_number, levels = c('0','1','2'))) %>%  -->
<!--   separate(sample, into=c('sample','time'), sep = '-') %>%  -->
<!--   mutate(arcas_pred_0 = ifelse(arcas_copy_number == 0, 1, 0)) %>%  -->
<!--   mutate(arcas_pred_1 = ifelse(arcas_copy_number == 1, 1, 0)) %>%  -->
<!--   mutate(arcas_pred_2 = ifelse(arcas_copy_number == 2, 1, 0)) -->
<!-- arcas_copy_df -->
<!-- ``` -->

<!-- ```{r} -->
<!-- drb_outcome_2 <- drb_outcome %>%  -->
<!--   left_join(arcas_copy_df, by = c('sample','time')) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- drb_outcome_2 %>%  -->
<!--   roc_curve(truth = copy_number, arcas_pred_0:arcas_pred_2) %>%  -->
<!--   autoplot() -->
<!-- drb_outcome_2 %>%  -->
<!--   roc_auc(truth = copy_number, arcas_pred_0:arcas_pred_2) -->


<!-- conf_mat(drb_outcome_2, -->
<!--          truth = copy_number, -->
<!--          estimate = arcas_copy_number) %>%  -->
<!--   autoplot(type = 'mosaic') -->
<!-- ``` -->

<!-- ### Accuracy of baseline arcas -->
<!-- ```{r} -->
<!-- drb_outcome -->
<!-- ``` -->
<!-- ```{r} -->
<!-- arcas_copy_df <- all_hla_expanded %>% filter(genotyper == "arcasHLA" & grepl("DRB[345]", locus)) %>%  -->
<!--   ungroup() %>%  -->
<!--   count(sample,locus, name = "arcas_copy_number") %>%  -->
<!--   mutate(arcas_copy_number = factor(arcas_copy_number, levels = c('0','1','2'))) %>%  -->
<!--   separate(sample, into=c('sample','time'), sep = '-') %>%  -->
<!--   mutate(arcas_pred_0 = ifelse(arcas_copy_number == 0, 1, 0)) %>%  -->
<!--   mutate(arcas_pred_1 = ifelse(arcas_copy_number == 1, 1, 0)) %>%  -->
<!--   mutate(arcas_pred_2 = ifelse(arcas_copy_number == 2, 1, 0)) -->
<!-- arcas_copy_df -->
<!-- ``` -->

<!-- ```{r} -->
<!-- drb_outcome_2 <- drb_outcome %>%  -->
<!--   left_join(arcas_copy_df, by = c('sample','time')) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- drb_outcome_2 %>%  -->
<!--   roc_curve(truth = copy_number, arcas_pred_0:arcas_pred_2) %>%  -->
<!--   autoplot() -->
<!-- drb_outcome_2 %>%  -->
<!--   roc_auc(truth = copy_number, arcas_pred_0:arcas_pred_2) -->


<!-- conf_mat(drb_outcome_2, -->
<!--          truth = copy_number, -->
<!--          estimate = arcas_copy_number) %>%  -->
<!--   autoplot(type = 'mosaic') -->
<!-- ``` -->

