---
title: "6) Accuracy on 3' 10X platform and bulk"
output: html_notebook
---

# TO DO
### ADD ALL SCRNA SAMPLES
  - Currently includes only those found in both scRNA and bulk RNA
### CORRECT combine_HLA_import for invitro arguments?
### UPDATE DRB4 PREDICTIONS - complete?

# DONE  
### Extend sample C in vitro typing to Ck, C1, and C2 predicted typing
### Need to correctly run HLAminer for paired end reads in bulk
  - Successfully fixed HPRA
  - Problems with implementing HPTASR (perl modules). However, this can't be used for SE reads, so not fully relevant to the paper and others have benchmarked HPRA vs. HPTASR for bulk RNA elsewhere 

# Reference info

- Publication: https://pubmed.ncbi.nlm.nih.gov/30518681/
- Ck and Sk represent a second time point for samples C and S, respectively
  - *"A second bone marrow aspiration was performed for 2 donors (Ck, Sk) (biological replicates) either 2 or 5 months after their first aspiration, respectively."*
- C1 and C2 represent the same sample split, technical replicates
  - From: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3396165
  - *"Description 	Technical replicate of C1"*


# Setup
```{r, message=F}
library(tidyverse)
library(tidymodels)
library(cowplot)
source("data_import_functions.R")
source("calculation_functions.R")
source("figure_format_functions.R")
isb_path <- "/labs/khatrilab/solomonb/covid/isb"
isb_samples <- read_tsv("/labs/khatrilab/solomonb/covid/isb/logs/210217_232725/parallel.log") %>% 
  separate(Command, into = c(NA, "sample"), sep = " ") %>% 
  pull(sample) %>% unique()
```

```{r}
invitro_hla <- invitro_import(path = "pmid30518681_scisco_hla.tsv", exclude_comment_samples = T) %>% 
  mutate(sample = gsub("STU-HD-","",sample))

invitro_hla_c <- invitro_hla %>% filter(sample == "C")

invitro_hla <- bind_rows(
  invitro_hla,
  invitro_hla_c %>% mutate(sample = "Ck"),
  invitro_hla_c %>% mutate(sample = "C1"),
  invitro_hla_c %>% mutate(sample = "C2")
)
```
# 3'
### Import
```{r}
# read_lines("/labs/khatrilab/solomonb/covid/pmid30518681/scRNA/SRR_Acc_List.txt")
pmid_hla <- combine_HLA_import(
  path = "/labs/khatrilab/solomonb/covid/pmid30518681/scRNA/",
  samples = read_lines("/labs/khatrilab/solomonb/covid/pmid30518681/scRNA/SRR_Acc_List.txt"),
  invitro_path = "pmid30518681_scisco_hla.tsv",
  filter_invitro = T
) %>% 
  rename(original_name = sample) %>% 
  mutate(original_name = gsub("^STU-HD-","",original_name)) %>% 
  # mutate(sra = ifelse(grepl("^SRR", original_name), original_name, NA)) %>% 
  left_join(
    read_csv("/labs/khatrilab/solomonb/covid/pmid30518681/pmid30518681_samples.csv"),
    by = c("original_name" = "Run")
  ) %>% 
  mutate(sample = ifelse(is.na(sample), original_name, sample)) %>% 
  select(-original_name, -platform) %>% 
  bind_rows(invitro_hla) 
pmid_hla <- format_hla_table(pmid_hla) %>% 
  select(sample,allele_id,locus,field_1,field_2,field_3,genotyper)
pmid_hla
pmid_hla %>% count(sample, genotyper)
```

### Prediction frequency
```{r}
pmid_arcas <- pmid_hla %>% filter(genotyper == "arcasHLA") %>% pull(sample) %>% unique()

prediction_frequency <- pmid_hla %>% 
  filter(sample %in% pmid_arcas) %>% 
  select(-allele_id) %>% 
  pivot_longer(cols = field_1:field_3, names_to = "field", values_to = "allele") %>% 
  pivot_wider(names_from = "genotyper", values_from = "allele", values_fn = function(x) sum(!is.na(x)), values_fill = 0) %>% 
  # filter(invitro != 0) %>% 
  ### NEED TO DEAL WITH HLAMINER AMBIGOUS ALLELES
  # mutate_at(vars(arcasHLA:hlaminer), funs(./invitro)) %>% 
  mutate_at(vars(arcasHLA:hlaminer), funs(./2)) %>% 
  select(-invitro) %>%
  ungroup() %>% 
  pivot_longer(cols = arcasHLA:hlaminer, names_to = "genotyper", values_to = "frequency")
```

```{r, fig.width = 5, fig.height = 8}
predictions_scRNA_plt <- gg_hla_prediction_frequency(prediction_frequency)
predictions_scRNA_plt
```

### Success
```{r}
success_df <- compare_hla(hla_df = pmid_hla %>% filter(sample %in% pmid_arcas), 
            reference = "invitro", 
            exclude_missing = T, 
            compare_function = "accuracy") 
```

```{r, fig.width = 5, fig.height = 4}
success_scRNA_plt <- calculate_summary_df(success_df) %>% 
  gg_summary_hla_accuracy(color_label = "Success")
success_scRNA_plt
```

### Accuracy




```{r}
accuracy_df <- compare_hla(hla_df = pmid_hla %>% filter(sample %in% pmid_arcas), 
            reference = "invitro", 
            exclude_missing = F, 
            compare_function = "accuracy") 
```

```{r, fig.width = 5, fig.height = 4}
accuracy_scRNA_plt <- calculate_summary_df(accuracy_df) %>% 
  gg_summary_hla_accuracy(color_label = "Accuracy")
accuracy_scRNA_plt
```

# Bulk

### Import
```{r}


# read_lines("/labs/khatrilab/solomonb/covid/pmid30518681/scRNA/SRR_Acc_List.txt")
pmid_bulk_hla <- combine_HLA_import(
  path = "/labs/khatrilab/solomonb/covid/pmid30518681/bulkRNA",
  samples = read_lines("/labs/khatrilab/solomonb/covid/pmid30518681/bulkRNA/bulk_sra_accession.txt"),
  invitro_path = "pmid30518681_scisco_hla.tsv",
  filter_invitro = T
) %>% 
  rename(original_name = sample) %>% 
  # mutate(original_name = gsub("^STU-HD-","",original_name)) %>% 
  # mutate(sra = ifelse(grepl("^SRR", original_name), original_name, NA)) %>% 
  left_join(
    read_csv("/labs/khatrilab/solomonb/covid/pmid30518681/pmid30518681_samples.csv"),
    by = c("original_name" = "Run")
  ) %>% 
  mutate(sample = ifelse(is.na(sample), original_name, sample)) %>% 
  select(-original_name, -platform) %>% 
  bind_rows(invitro_hla)
pmid_bulk_hla <- format_hla_table(pmid_bulk_hla) %>% 
  select(sample,allele_id,locus,field_1,field_2,field_3,genotyper)
pmid_bulk_hla
```
```{r}
bulk_samples <- pmid_bulk_hla %>% 
  filter(genotyper != "invitro") %>% 
  count(sample) %>% 
  pull(sample)
```

### Prediction frequency
```{r}
prediction_frequency <- pmid_bulk_hla %>% 
  filter(sample %in% bulk_samples) %>% 
  select(-allele_id) %>% 
  pivot_longer(cols = field_1:field_3, names_to = "field", values_to = "allele") %>% 
  pivot_wider(names_from = "genotyper", values_from = "allele", values_fn = function(x) sum(!is.na(x)), values_fill = 0) %>% 
  # filter(invitro != 0) %>% 
  ### NEED TO DEAL WITH HLAMINER AMBIGOUS ALLELES
  # mutate_at(vars(arcasHLA:hlaminer), funs(./invitro)) %>% 
  mutate_at(vars(arcasHLA:hlaminer), funs(./2)) %>% 
  select(-invitro) %>%
  ungroup() %>% 
  pivot_longer(cols = arcasHLA:hlaminer, names_to = "genotyper", values_to = "frequency")
```

```{r, fig.width = 5, fig.height = 8}
predictions_bulkRNA_plt <- gg_hla_prediction_frequency(prediction_frequency)
predictions_bulkRNA_plt
```

### Success
```{r}
success_df <- compare_hla(hla_df = pmid_bulk_hla %>% filter(sample %in% bulk_samples), 
            reference = "invitro", 
            exclude_missing = T, 
            compare_function = "accuracy") 
```

```{r, fig.width = 5, fig.height = 4}
success_bulkRNA_plt <- calculate_summary_df(success_df) %>% 
  gg_summary_hla_accuracy(color_label = "Success")
success_bulkRNA_plt
```

```{r}
accuracy_df <- compare_hla(hla_df = pmid_bulk_hla %>% filter(sample %in% bulk_samples), 
            reference = "invitro", 
            exclude_missing = F, 
            compare_function = "accuracy") 
```

```{r, fig.width = 5, fig.height = 4}
accuracy_bulkRNA_plt <- calculate_summary_df(accuracy_df) %>% 
  gg_summary_hla_accuracy(color_label = "Accuracy")
accuracy_bulkRNA_plt
```
# Combined plot

```{r,fig.height = 6, fig.width = 16}
combine_plt <- cowplot::plot_grid(
  cowplot::plot_grid(
    predictions_scRNA_plt,
    cowplot::plot_grid(
      success_scRNA_plt,
      accuracy_scRNA_plt,
      ncol=1
    ),
    nrow=1
  ),
  cowplot::plot_grid(
    predictions_bulkRNA_plt,
    cowplot::plot_grid(
      success_bulkRNA_plt,
      accuracy_bulkRNA_plt,
      ncol=1
    ),
    nrow=1
  ),
  nrow=1
)

cowplot::plot_grid(
  cowplot::plot_grid(
    NULL,
    NULL,
    nrow=1,
    labels = c("3' scRNA sequencing", "Bulk RNA sequencing")
  ),
  combine_plt,
  ncol=1,
  rel_heights = c(0.05,0.95)
)
```



# DRB345 copy number modeling

### Calculations
```{r}
### UPDATED FROM data_import_functions.R to use direct arcas log raether than pipeline arcas log
### Negates the issue of identifying the write pipeline run
# Import HLA loci alignment stats from arcas logs
hla_mapping_stats_import <- function(samples, log_dir){
  tibble(sample = samples) %>% 
    mutate(data = map(sample, function(x){
      log_path <- sprintf("%s/%s.genotype.log",log_dir,x)
      df <- tibble(lines = read_lines(log_path))
      if (any(grepl("error", df$lines, ignore.case = T))){
        NA
      } else {
        df %>% 
          mutate(lines = gsub("\t", "", lines)) %>% 
          filter(grepl("^HLA", lines)) %>% 
          separate(lines, into = c("locus", "abundance", "reads", "classes"), sep = " +")
      }
    })) %>%
    unnest(data) %>% 
    select(!(contains("data"))) %>% 
    separate(locus, into = c(NA, "locus"), sep = "-")
}
```



```{r}
arcas_log_3p_dir <- "/labs/khatrilab/solomonb/covid/pmid30518681/scRNA/log"
arcas_log_bulk_dir <- "/labs/khatrilab/solomonb/covid/pmid30518681/bulkRNA/arcasHLA"

samples_3p <- read_tsv("/labs/khatrilab/solomonb/covid/pmid30518681/scRNA/SRR_Acc_List.txt", col_names = "srr") %>% pull(srr)
samples_bulk <- read_tsv("/labs/khatrilab/solomonb/covid/pmid30518681/bulkRNA/bulk_sra_accession.txt", col_names = "srr") %>% pull(srr)

alignment_stats_3p_df <- hla_mapping_stats_import(samples_3p, arcas_log_3p_dir)
alignment_stats_bulk_df <- hla_mapping_stats_import(samples_bulk, arcas_log_bulk_dir)
```

```{r}
# Get DRB345/DRB1 ratios
drb345_ratios_3p <- get_drb345_ratios(alignment_stats_3p_df)
drb345_ratios_bulk <- get_drb345_ratios(alignment_stats_bulk_df)

# Rename pmid samples from SRA to sample names
drb345_ratios_3p <- pmid_sample_rename(drb345_ratios_3p)
drb345_ratios_bulk <- pmid_sample_rename(drb345_ratios_bulk)
```

```{r}
pmid_copy_number <- get_drb345_copy_number("pmid30518681_scisco_hla.tsv") %>% 
  mutate(sample = gsub(".*HD-","",sample))

# Duplicate in vitro typing for replicate samples Ck, C1, C2
pmid_copy_number_c <- pmid_copy_number %>% filter(sample == "C")

pmid_copy_number <- bind_rows(
  pmid_copy_number,
  pmid_copy_number_c %>% mutate(sample = "Ck"),
  pmid_copy_number_c %>% mutate(sample = "C1"),
  pmid_copy_number_c %>% mutate(sample = "C2")
)
```

```{r}
# Combine DRB345 sequencing ratios and ground truth copy numbers
drb345_df_3p <- create_drb345_df(drb345_ratios_3p, pmid_copy_number)
drb345_df_bulk <- create_drb345_df(drb345_ratios_bulk, pmid_copy_number)
```



```{r}
# Apply KNN model
drb_fit <- readRDS("drb345_knn_model.RDS")
drb345_3p_outcome <- drb345_df_3p %>%
  bind_cols(
    drb_fit %>% predict(new_data = drb345_df_3p),
    drb_fit %>% predict(new_data = drb345_df_3p, type = 'prob')
  ) %>% 
  mutate(copy_number = factor(copy_number))

drb345_bulk_outcome <- drb345_df_bulk %>%
  bind_cols(
    drb_fit %>% predict(new_data = drb345_df_bulk),
    drb_fit %>% predict(new_data = drb345_df_bulk, type = 'prob')
  )%>% 
  mutate(copy_number = factor(copy_number, levels = c("0","1", "2")))
```

### ROC plots and AUC calculations
```{r}
# Generate ROC plots and AUC
drb345_ROC_3p <- gg_multilevel_roc(drb345_3p_outcome)
drb345_ROC_3p
drb345_3p_outcome %>% roc_auc(truth = copy_number, .pred_0:.pred_2)


drb345_ROC_bulk <- drb345_bulk_outcome %>% 
  # None of the bulk RNA samples have 2 alleles for a single DRB345. 
  # ROC will fail to generate without all possible allele counts {0,1,2} in ground truth
  # Will add in place holder {2}, which will allow ROC to proceed
  # Will exclude class-{2} ROC plot
  # class-{0} and class-{1} plots are not affected by the "pseudotruth" addition
  rbind(., list("2_copy_pseudotruth",0,0,0,"DRB3","2","2",0,0,1)) %>% 
  # Proceed normally from here
  roc_curve(truth = copy_number, .pred_0:.pred_2) %>% 
  autoplot() %>% 
  (function(x) ggplot_build(x)$data[[1]]) %>% 
  mutate(n_alleles = fct_recode(PANEL, "0"="1", "1"="2", "2"="3")) %>% 
  filter(n_alleles != "2") %>% # Removing {2} pseudotruth here
  ggplot(aes(x=x,y=y,color=n_alleles))+
  geom_path()+
  geom_abline(slope=1,intercept=0, linetype="dotted")+
  theme_bw() +
  scale_color_brewer(palette = "Set1") +
  labs(x = "1 - Specificity", y = "Sensitivity", color = "Number\nof alleles")
drb345_ROC_bulk
drb345_bulk_outcome %>% roc_auc(truth = copy_number, .pred_0:.pred_2)
```

```{r}
# Demonstrate that pseudotruth row doesn't affect other ROC classes
# Reformats outcome as a binary classifier. 
# Event_level specifies {1} outcome as the base event, so plot should match {1} plot above
drb345_bulk_outcome %>% 
  mutate(copy_number = factor(copy_number, levels = c("0","1"))) %>% 
  roc_curve(truth = copy_number, .pred_1, event_level = "second") %>% 
  autoplot()
```
### DRB345 ratio distribution plots

```{r}
drb345_df_5p <- read_csv("isb_drb345_df.csv")
drb345_ratio_plt_3p <- gg_drb345_ratio(drb345_df_3p, drb345_df_5p)
drb345_ratio_plt_bulk <- gg_drb345_ratio(drb345_df_bulk, drb345_df_5p)
drb345_ratio_plt_3p
drb345_ratio_plt_bulk
```
### Combined DRB345 modeling plot

```{r, fig.width=10, fig.height=8}
plot_grid(
  drb345_ROC_3p,
  drb345_ratio_plt_3p,
  drb345_ROC_bulk,
  drb345_ratio_plt_bulk,
  nrow = 2,
  ncol = 2,
  labels = LETTERS[1:4],
  align = "h",
  axis = "bt"
)
```

# Update DRB345 alleles based on copy number modeling



































# OLD WORK
<!-- ```{r, fig.width = 8, fig.height = 4} -->
<!-- c1_a <- accuracy_df%>%  -->
<!--   filter(grepl("^[ABC]", locus)) %>%  -->
<!--   gg_accuracy() -->

<!-- dp_a <- accuracy_df %>%  -->
<!--   filter(grepl("^DP", locus) & genotyper %in% c("arcasHLA", "hlaminer"),) %>%  -->
<!--   gg_accuracy() -->

<!-- dq_a <- accuracy_df %>%  -->
<!--   filter(grepl("^DQ", locus) & genotyper %in% c("arcasHLA", "hlaminer", "phlat"),) %>%  -->
<!--   gg_accuracy() -->

<!-- drb1_a <- accuracy_df %>%  -->
<!--   filter(grepl("^DRB1", locus) & genotyper %in% c("arcasHLA", "hlaminer", "phlat"),) %>%  -->
<!--   gg_accuracy() -->

<!-- drb345_a <- accuracy_df %>%  -->
<!--   filter(grepl("^DRB[345]", locus) & genotyper %in% c("arcasHLA", "hlaminer"),) %>%  -->
<!--   gg_accuracy() -->

<!-- c2_a <- dp_a| -->
<!--   strip_y(dq_a)| -->
<!--   strip_y(drb1_a)| -->
<!--   strip_y(drb345_a)| -->
<!--   plot_layout(widths = c(6,6,3,9)) -->

<!-- c1_a | c2_a | plot_layout(widths = c(4,13)) -->
<!-- ``` -->



