---
title: "6) Accuracy on 3' 10X platform and bulk"
output: html_notebook
---

# TO DO
### ADD ALL SCRNA SAMPLES
  - Currently includes only those found in both scRNA and bulk RNA

  
### CORRECT combine_HLA_import for invitro arguments
### UPDATE DRB4 PREDICTIONS - complete?
### Need to correctly run HLAminer for paired end reads in bulk
  - Successfully fixed HPRA
  - Problems with implementing HPTASR (perl modules). However, this can't be used for SE reads, so not fully relevant to the paper and others have benchmarked HPRA vs. HPTASR for bulk RNA elsewhere 
### Extend sample C in vitro typing to Ck, C1, and C2 predicted typing

# Reference info

- Publication: https://pubmed.ncbi.nlm.nih.gov/30518681/
- Ck and Sk represent a second time point for samples C and S, respectively
  - *"A second bone marrow aspiration was performed for 2 donors (Ck, Sk) (biological replicates) either 2 or 5 months after their first aspiration, respectively."*
- C1 and C2 represent the same sample split, technical replicates
  - From: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3396165
  - *"Description 	Technical replicate of C1"*


# Setup
```{r, message=F}
library(tidyverse)
library(ggh4x)
library(patchwork)
source("data_import_functions.R")
source("accuracy_functions.R")
source("ggplot_layouts.R")
isb_path <- "/labs/khatrilab/solomonb/covid/isb"
isb_samples <- read_tsv("/labs/khatrilab/solomonb/covid/isb/logs/210217_232725/parallel.log") %>% 
  separate(Command, into = c(NA, "sample"), sep = " ") %>% 
  pull(sample) %>% unique()
```

```{r}
invitro_hla <- invitro_import(path = "pmid30518681_scisco_hla.tsv", exclude_comment_samples = T) %>% 
  mutate(sample = gsub("STU-HD-","",sample))

invitro_hla_c <- invitro_hla %>% filter(sample == "C")

invitro_hla <- bind_rows(
  invitro_hla,
  invitro_hla_c %>% mutate(sample = "Ck"),
  invitro_hla_c %>% mutate(sample = "C1"),
  invitro_hla_c %>% mutate(sample = "C2")
)
```
# 3'
### Import
```{r}
# read_lines("/labs/khatrilab/solomonb/covid/pmid30518681/scRNA/SRR_Acc_List.txt")
pmid_hla <- combine_HLA_import(
  path = "/labs/khatrilab/solomonb/covid/pmid30518681/scRNA/",
  samples = read_lines("/labs/khatrilab/solomonb/covid/pmid30518681/scRNA/SRR_Acc_List.txt"),
  invitro_path = "pmid30518681_scisco_hla.tsv",
  filter_invitro = T
) %>% 
  rename(original_name = sample) %>% 
  mutate(original_name = gsub("^STU-HD-","",original_name)) %>% 
  # mutate(sra = ifelse(grepl("^SRR", original_name), original_name, NA)) %>% 
  left_join(
    read_csv("/labs/khatrilab/solomonb/covid/pmid30518681/pmid30518681_samples.csv"),
    by = c("original_name" = "Run")
  ) %>% 
  mutate(sample = ifelse(is.na(sample), original_name, sample)) %>% 
  select(-original_name, -platform) %>% 
  bind_rows(invitro_hla) 
pmid_hla <- format_hla_table(pmid_hla) %>% 
  select(sample,allele_id,locus,field_1,field_2,field_3,genotyper)
pmid_hla
pmid_hla %>% count(sample, genotyper)
```

### Prediction frequency
```{r}
pmid_arcas <- pmid_hla %>% filter(genotyper == "arcasHLA") %>% pull(sample) %>% unique()

prediction_frequency <- pmid_hla %>% 
  filter(sample %in% pmid_arcas) %>% 
  select(-allele_id) %>% 
  pivot_longer(cols = field_1:field_3, names_to = "field", values_to = "allele") %>% 
  pivot_wider(names_from = "genotyper", values_from = "allele", values_fn = function(x) sum(!is.na(x)), values_fill = 0) %>% 
  # filter(invitro != 0) %>% 
  ### NEED TO DEAL WITH HLAMINER AMBIGOUS ALLELES
  # mutate_at(vars(arcasHLA:hlaminer), funs(./invitro)) %>% 
  mutate_at(vars(arcasHLA:hlaminer), funs(./2)) %>% 
  select(-invitro) %>%
  ungroup() %>% 
  pivot_longer(cols = arcasHLA:hlaminer, names_to = "genotyper", values_to = "frequency")
```

```{r, fig.width = 5, fig.height = 8}
predictions_plt <- gg_hla_prediction_frequency(prediction_frequency)
predictions_plt
```

### Success
```{r}
success_df <- compare_hla(hla_df = pmid_hla %>% filter(sample %in% pmid_arcas), 
            reference = "invitro", 
            exclude_missing = T, 
            compare_function = "accuracy") 
```

```{r, fig.width = 5, fig.height = 4}
success_plt <- calculate_summary_df(success_df) %>% 
  gg_summary_hla_accuracy(color_label = "Success")
success_plt
```

### Accuracy




```{r}
accuracy_df <- compare_hla(hla_df = pmid_hla %>% filter(sample %in% pmid_arcas), 
            reference = "invitro", 
            exclude_missing = F, 
            compare_function = "accuracy") 
```

```{r, fig.width = 5, fig.height = 4}
accuracy_plt <- calculate_summary_df(accuracy_df) %>% 
  gg_summary_hla_accuracy(color_label = "Accuracy")
accuracy_plt
```

# Bulk

### Import
```{r}


# read_lines("/labs/khatrilab/solomonb/covid/pmid30518681/scRNA/SRR_Acc_List.txt")
pmid_bulk_hla <- combine_HLA_import(
  path = "/labs/khatrilab/solomonb/covid/pmid30518681/bulkRNA",
  samples = read_lines("/labs/khatrilab/solomonb/covid/pmid30518681/bulkRNA/bulk_sra_accession.txt"),
  invitro_path = "pmid30518681_scisco_hla.tsv",
  filter_invitro = T
) %>% 
  rename(original_name = sample) %>% 
  # mutate(original_name = gsub("^STU-HD-","",original_name)) %>% 
  # mutate(sra = ifelse(grepl("^SRR", original_name), original_name, NA)) %>% 
  left_join(
    read_csv("/labs/khatrilab/solomonb/covid/pmid30518681/pmid30518681_samples.csv"),
    by = c("original_name" = "Run")
  ) %>% 
  mutate(sample = ifelse(is.na(sample), original_name, sample)) %>% 
  select(-original_name, -platform) %>% 
  bind_rows(invitro_hla)
pmid_bulk_hla <- format_hla_table(pmid_bulk_hla) %>% 
  select(sample,allele_id,locus,field_1,field_2,field_3,genotyper)
pmid_bulk_hla
```
```{r}
bulk_samples <- pmid_bulk_hla %>% 
  filter(genotyper != "invitro") %>% 
  count(sample) %>% 
  pull(sample)
```


### Success
```{r}
success_df <- compare_hla(hla_df = pmid_bulk_hla %>% filter(sample %in% bulk_samples), 
            reference = "invitro", 
            exclude_missing = T, 
            compare_function = "accuracy") 
```

```{r, fig.width = 5, fig.height = 4}
success_plt <- calculate_summary_df(success_df) %>% 
  gg_summary_hla_accuracy(color_label = "Success")
success_plt
```

```{r}
accuracy_df <- compare_hla(hla_df = pmid_bulk_hla %>% filter(sample %in% bulk_samples), 
            reference = "invitro", 
            exclude_missing = F, 
            compare_function = "accuracy") 
```

```{r, fig.width = 5, fig.height = 4}
accuracy_plt <- calculate_summary_df(accuracy_df) %>% 
  gg_summary_hla_accuracy(color_label = "Accuracy")
accuracy_plt
```

```{r, fig.width = 8, fig.height = 4}
c1_a <- accuracy_df%>% 
  filter(grepl("^[ABC]", locus)) %>% 
  gg_accuracy()

dp_a <- accuracy_df %>% 
  filter(grepl("^DP", locus) & genotyper %in% c("arcasHLA", "hlaminer"),) %>% 
  gg_accuracy()

dq_a <- accuracy_df %>% 
  filter(grepl("^DQ", locus) & genotyper %in% c("arcasHLA", "hlaminer", "phlat"),) %>% 
  gg_accuracy()

drb1_a <- accuracy_df %>% 
  filter(grepl("^DRB1", locus) & genotyper %in% c("arcasHLA", "hlaminer", "phlat"),) %>% 
  gg_accuracy()

drb345_a <- accuracy_df %>% 
  filter(grepl("^DRB[345]", locus) & genotyper %in% c("arcasHLA", "hlaminer"),) %>% 
  gg_accuracy()

c2_a <- dp_a|
  strip_y(dq_a)|
  strip_y(drb1_a)|
  strip_y(drb345_a)|
  plot_layout(widths = c(6,6,3,9))

c1_a | c2_a | plot_layout(widths = c(4,13))
```



