---
title: "R Notebook"
output: html_notebook
---

```{r}
library(tidyverse)
source("data_import_functions.R")
all_hla_expanded <- readRDS("all_hla_expanded.RDS")
```


# Running scHLA count

```{r}
hla_samples <- read_tsv("../../covid/isb/scHLAcount/all_fastq_files.txt", col_names = "file") %>% 
  filter(grepl("^INCOV", file)) %>% 
  filter(grepl("-BL", file)) %>% 
  separate(file, into = c("sample", NA), sep = "_", remove = F) %>% 
  drop_na()
hla_samples
```

```{r}
select_last_allele <- function(x){
  x[!is.na(x)] %>%
    tail(1) %>% 
    str_replace_all("_", ":")}
hla_key <- all_hla_expanded %>% 
  rowwise() %>% 
  filter(grepl("^[ABC]", locus)) %>% 
  mutate(allele = select_last_allele(across(contains("field")))) %>% 
  select(sample, genotyper, locus, allele) %>% 
  unite(allele, locus, allele, sep = "*")
hla_key
```

```{r}
hla_merge <- hla_samples %>% 
  left_join(hla_key, by = "sample") %>% 
  drop_na() %>% 
  select(-sample) %>% 
  group_by(file, genotyper) %>% 
  nest()
hla_merge
```

```{r}
hla_merge %>% 
  # head() %>% 
  mutate(write = pmap(list(data, file, genotyper), function(d,f,g){
    dir <- sprintf("../../covid/isb/scHLAcount/genotypes/%s",g)
    if (!dir.exists(dir)){dir.create(dir, recursive = T)}
    write_tsv(d, 
              sprintf("%s/%s_hla.tsv",dir,f), 
              col_names = F,
              )}))
```

### Run script

`sbatch /covid/scripts/isb_scHLAcount_benchmark.sh`

# UMAP projections

```{r}
srt <- readRDS("/labs/khatrilab/hongzheng/webapps/isb/srt_isb260.meta.rds")
```

```{r}
srt %>% 
  filter(sampleID == "INCOV005-BL") %>% 
  # sample_frac(0.1) %>%
  ggplot(aes(x = UMAP_1, y = UMAP_2, color = celltype)) +
  geom_point(size = 0.5, alpha = 0.5)+
  theme_bw() 

```


```{r}
x <- scHLA_data_processing(
  sample="INCOV005-BL_S7",
  result_dir=sprintf("%s/scHLAcount/output/invitro", isb_path),
  barcode_dir=sprintf("%s/scHLAcount/barcodes", isb_path)
)
x
```

```{r}
y <- x %>% select(cell, allele, allele_order, locus, allele_ratio) %>% 
  pivot_wider(names_from = "allele_order", values_from = c("allele", "allele_ratio")) %>% 
  filter(locus == "A")
```

```{r}
srt %>% 
  filter(sampleID == "INCOV005-BL") %>% 
  left_join(x %>% filter(locus == "A"), by = "cell") %>% 
  filter(!is.na(allele)) %>% 
    ggplot(aes(x = UMAP_1, y = UMAP_2, color = allele_ratio)) +
  geom_point(size = 0.5, alpha = 0.5)+
  theme_bw() +
  facet_grid(.~allele)+
  scale_color_gradient2(high = "red", mid = "grey80", low = "blue", midpoint = 0.5, na.value = "transparent")
```

```{r}

```

```{r}

```

```{r, fig.width = 9, fig.height = 8}
samp <- "INCOV008-BL"
scHLAcount_dir <- sprintf("%s/scHLAcount", isb_path)

# Create tibble of all genotyper and sample combinations
allele_data <- tibble(genotyper = c("invitro", "arcasHLA", "optitype", "phlat", "hlaminer")) %>% 
  mutate(data = map(genotyper, function(x) suppressMessages(read_tsv("../../covid/isb/scHLAcount/BL_fastq_files.txt", col_names = "sample")))) %>% 
  unnest(data) %>% 
  filter(grepl(samp, sample)) %>% 
  # Import data based on sample and genotyper
  mutate(result_path = sprintf("%s/output/%s",scHLAcount_dir, genotyper),
         barcode_path = sprintf("%s/barcodes", scHLAcount_dir)) %>% 
  # head(2) %>%
  mutate(data = pmap(list(sample, result_path, barcode_path), function(s,r,b){
    scHLA_data_processing(
      sample=s,
      result_dir=r,
      barcode_dir=b
    )
  })) %>% unnest(data) 

srt %>% 
  filter(sampleID == samp) %>% 
  left_join(allele_data %>% filter(locus == "A"), by = "cell") %>% 
  filter(!is.na(allele)) %>% 
    ggplot(aes(x = UMAP_1, y = UMAP_2, color = allele_ratio)) +
  geom_point(size = 0.5, alpha = 0.5)+
  theme_bw() +
  facet_grid(genotyper~allele)+
  scale_color_gradient2(high = "red", mid = "grey80", low = "blue", midpoint = 0.5, na.value = "transparent")
```
```{r, fig.width = 9, fig.height = 4}
samp <- "INCOV007-BL"
scHLAcount_dir <- sprintf("%s/scHLAcount", isb_path)

# Create tibble of all genotyper and sample combinations
allele_data <- tibble(genotyper = c("invitro", "arcasHLA", "optitype", "phlat", "hlaminer")) %>% 
  mutate(data = map(genotyper, function(x) suppressMessages(read_tsv("../../covid/isb/scHLAcount/BL_fastq_files.txt", col_names = "sample")))) %>% 
  unnest(data) %>% 
  filter(grepl(samp, sample)) %>% 
  # Import data based on sample and genotyper
  mutate(result_path = sprintf("%s/output/%s",scHLAcount_dir, genotyper),
         barcode_path = sprintf("%s/barcodes", scHLAcount_dir)) %>% 
  # head(2) %>%
  mutate(data = pmap(list(sample, result_path, barcode_path), function(s,r,b){
    scHLA_data_processing(
      sample=s,
      result_dir=r,
      barcode_dir=b
    )
  })) %>% unnest(data) %>% 
  select(genotyper, cell, locus, allele_order, allele_ratio) %>% 
  # group_by(genotyper, cell, allele_order) %>%  mutate(test = length(allele_order))
  pivot_wider(names_from = "allele_order", values_from = "allele_ratio", names_prefix = "allele_") %>% 
  rowwise() %>% 
  mutate(max_ratio = max(across(contains("allele_")), na.rm = T)) %>% 
  select(genotyper, cell, locus, max_ratio)


srt %>% 
  filter(sampleID == samp) %>% 
  left_join(allele_data %>% filter(locus == "A"), by = "cell") %>% 
  mutate(genotyper = fct_relevel(genotyper, "invitro")) %>% 
  filter(!is.na(genotyper)) %>%
    ggplot(aes(x = UMAP_1, y = UMAP_2, color = max_ratio)) +
  geom_point(size = 0.5, alpha = 0.5)+
  theme_bw() +
  facet_grid(.~genotyper)+
  scale_color_gradient(high = "red", low = "blue", na.value = "transparent")


# WHERE IS OPTITYPE
```