---
title: "R Notebook"
output: html_notebook
---

```{r}
library(tidyverse)
library(ggcorrplot)
library(cowplot)
library(ggpubr)
```
# All sample heatmap
```{r, fig.width = 16, fig.height = 2}
for (l in c("A", "B", "C")){
  plt <- accuracy_df %>% 
    ungroup() %>% 
    filter(locus == l & grepl("BL$", sample) & field == "field_2") %>% 
    select(sample, genotyper, accuracy) %>% 
    pivot_wider(names_from = "genotyper", values_from = "accuracy", values_fill = 0) %>% 
    column_to_rownames("sample") %>% 
    select(rev(c("arcasHLA", "optitype", "phlat", "hlaminer"))) %>% 
    # t() %>% 
    ggcorrplot::ggcorrplot(
      
    ) +
    scale_fill_gradientn(colours = viridis::viridis(3))
  assign(sprintf("%s_plot", l), plt)
}

```
# Summarized heatmap

```{r}
plt <- accuracy_df %>% 
  group_by(locus, field, genotyper) %>% 
  summarise(mean_accuracy = mean(accuracy),
            sd = sd(accuracy),
            se = sd(accuracy)/sqrt(n())) %>% 
  ungroup() %>% 
  mutate(mean_accuracy = ifelse(mean_accuracy == 0, NA, mean_accuracy), 
         sd = ifelse(mean_accuracy == 0, NA, sd),
         se = ifelse(mean_accuracy == 0, NA, se),
         locus = factor(locus, levels = sort(unique(locus), decreasing = T))) %>% 
  # select(-field) %>% 
  # pivot_wider(names_from = "genotyper", values_from = "accuracy") %>% 
  # column_to_rownames("locus") %>% 
  ggpubr::ggballoonplot(y = "locus", 
                x = "genotyper", 
                fill = "mean_accuracy", 
                size = "se", 
                facet.by = "field") +
  scale_fill_viridis_c(option = "cividis", limits = c(0,1))
# plt
# plt + 
#   theme_bw() +
#   theme(axis.text.x = element_text(angle = 90))+
#   labs(x = "", fill = "Accuracy", size = "SE") +
#   scale_size_continuous(breaks = c(0.9,0.8,0.7,0.6), 
#                         labels = rev(c(0.9,0.8,0.7,0.6)),
#                         range = c(0,10)
#                         ) +
#   ggtitle("4) Summarized heatmap")


points <- c(0,0.01,0.02,0.05,1)

plt + 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90))+
  labs(x = "", fill = "Accuracy", size = "SE") +
  scale_size_continuous(breaks = points, 
                        labels = rev(points),
                        range = c(1,10)
                        ) +
  ggtitle("4) Summarized heatmap")
```

```{r}
df <- accuracy_df %>% 
  group_by(locus, field, genotyper) %>% 
  summarise(mean_accuracy = mean(accuracy),
            sd = sd(accuracy),
            se = sd(accuracy)/sqrt(n())) %>% 
  ungroup() %>% 
  mutate(mean_accuracy = ifelse(mean_accuracy == 0, NA, mean_accuracy), 
         sd = ifelse(mean_accuracy == 0, NA, sd),
         se = ifelse(mean_accuracy == 0, NA, se),
         locus = factor(locus, levels = sort(unique(locus), decreasing = T))) %>% 
  mutate(class = ifelse(grepl("^[ABC]",locus),"mhc1","mhc2"))

size_scales <- c(0,0.02,0.04,0.06,0.08,0.1)
ggplot(df,aes(x = genotyper, y = locus)) +
  geom_point(aes(fill = mean_accuracy, size = 1-se), shape = 21, color = "black")+
  facet_grid(.~field)+
  theme_bw() +
  scale_fill_viridis_c(option = "plasma", limits = c(0,1)) +
  labs(x = "", y = "", fill = "Accuracy") +
  theme(axis.text.x = element_text(angle = 90)) +
  labs(x = "", fill = "Accuracy", size = "SE") +
  scale_size_continuous(breaks = 1-size_scales, 
                        labels = size_scales,
                        range = c(1,10)
                        )

ggplot(df,aes(x = genotyper, y = locus)) +
  geom_tile(aes(fill = mean_accuracy), color = "white")+
  # geom_text(aes(label = round(mean_accuracy, digits = 1)))+
  # ggrepel::geom_text_repel(aes(label = round(mean_accuracy, digits = 2)), 
  #                          size = 3.5, 
  #                          bg.color = "white",
  #                          force = 0,
  #                          force_pull = 0)+
  facet_grid(class~field, scales = "free_y", space = "free_y")+
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background.x = element_blank(),
        # strip.text.x = element_text(face = "bold"),
        strip.background.y = element_blank(),
        strip.text.y = element_blank()) +
  scale_fill_viridis_c(option = "plasma", limits = c(0,1), na.value = "transparent") +
  labs(x = "", y = "", fill = "Accuracy")

ggplot(df,aes(x = genotyper, y = locus)) +
  geom_tile(aes(fill = mean_accuracy), color = "white")+
  ggrepel::geom_text_repel(aes(label = round(mean_accuracy, digits = 2)),
                           size = 3.5,
                           bg.color = "white",
                           force = 0,
                           force_pull = 0)+
  facet_grid(class~field, scales = "free_y", space = "free_y")+
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background.x = element_blank(),
        # strip.text.x = element_text(face = "bold"),
        strip.background.y = element_blank(),
        strip.text.y = element_blank()) +
  scale_fill_viridis_c(option = "plasma", limits = c(0,1), na.value = "transparent") +
  labs(x = "", y = "", fill = "Accuracy")
```

```{r}
reformat_hla_field <- function(field_vector){
  suppressWarnings(
  field_vector %>% 
    factor() %>% 
    fct_recode(
      "Field 1" = "field_1",
      "Field 2" = "field_2",
      "Field 3" = "field_3"
    ) %>% 
    fct_relevel(
      "Field 1",
      "Field 2",
      "Field 3"
    ))
}

reformat_hla_field(df$field[1:5])
```
```{r}
reformat_hla_genotyper <- function(genotyper_vector){
  suppressWarnings(
  genotyper_vector %>% 
    factor() %>% 
    fct_recode(
      "HLAminer" = "hlaminer",
      "OptiType" = "optitype",
      "PHLAT" = "phlat"
    ) %>% 
    fct_relevel(
      "arcasHLA",
      "OptiType",
      "PHLAT",
      "HLAminer"
    ))
}

reformat_hla_genotyper(df$genotyper[1:10])
```
```{r, fig.width = 5, fig.height = 4}
for (color in c("C","D","E","F","G","H")){
  plt <- df %>% 
  mutate(field = reformat_hla_field(field)) %>% 
  mutate(genotyper = reformat_hla_genotyper(genotyper)) %>% 
  gg_hla_heatmap(color = color)
  print(plt)
}

```
### Tables
```{r, fig.width = 12}
library(flextable)
df %>% 
  mutate(field = reformat_hla_field(field)) %>% 
  mutate(genotyper = reformat_hla_genotyper(genotyper)) %>% 
  mutate(cell_value = sprintf("%s %s %s", round(mean_accuracy,2),"\u00B1",round(se,2) )) %>%
  mutate(cell_value = ifelse(grepl("NA", cell_value), NA, cell_value)) %>% 
  select(-sd,-se,-class, -mean_accuracy) %>% 
  pivot_wider(names_from = c("field","genotyper"), values_from = "cell_value", names_sep = "-") %>%
  {tibble(col_keys = names(.))->>test;.} %>% 
  flextable() %>% 
  colformat_char(na_str = "---") %>% 
  set_header_df(mapping = test %>%
                  separate(col_keys, into = c("field", "genotyper"), sep = "-", remove = F) %>% 
                  drop_na(),
                key = "col_keys") %>% 
  merge_h(part = "header") %>% 
  theme_vanilla() %>% 
  vline(j=c(1,5,9), border = fp_border_default()) %>% 
  fix_border_issues()
  
```



# Original plot

```{r, fig.height = 6, fig.width = 4}
accuracy_df %>% 
  filter(grepl("^[ABC]", locus)) %>% 
  mutate(genotyper = fct_relevel(genotyper, "arcasHLA", "phlat", "optitype", "hlaminer")) %>% 
    ggplot(aes(x = genotyper, y = accuracy))+
    stat_summary(fun = mean, geom = "point", position = position_dodge(width=0.9)) +
    stat_summary(fun.data = mean_se, geom = "errorbar", position=position_dodge(width=0.9), width = 0.5) +
    theme_bw() +
    # facet_nested(.~locus+field, scales = "free_y") +
    facet_grid(field~locus, scales = "free") +
    coord_cartesian(ylim = c(0,1))+
    scale_y_continuous(n.breaks = 6)+
    labs(y = "Accuracy", x = "")+
    theme(axis.text.x = element_text(angle = 90))+
    theme(panel.spacing = unit(0.1,"line")) +
  ggtitle("1) Original Plot")
```

# Barplot

```{r, fig.width = 12, fig.height = 4}
accuracy_df %>% 
  filter(grepl("^[ABC]", locus)) %>% 
  mutate(genotyper = fct_relevel(genotyper, "arcasHLA", "phlat", "optitype", "hlaminer")) %>% 
  ggplot(aes(x = field, y = accuracy, fill = genotyper))+
  stat_summary(fun = mean, geom = "bar", position = "dodge", color = "black")  +
  stat_summary(fun.data = mean_se, geom = "errorbar", position=position_dodge(width = 0.9), width = 0.5) +
  theme_bw()  +
  scale_fill_brewer(palette = "Set1") +
  facet_grid(.~locus) +
  labs(x = "") +
  ggtitle("2) Adjacent bar plot")
```

# Individual heatmap

```{r}
cell_key <- tibble(line = system("wc -l ~/solomonb/covid/isb/extracted_barcodes/INCOV*", intern = T)) %>% 
  separate(line, into = c("n_cells", "file"), sep = " /") %>% 
  drop_na() %>% 
  mutate(sample = gsub("\\..*","",basename(file))) %>% 
  select(-file)
```

# Import arcas logs
```{r}
alignment_stats <- tibble(sample = isb_samples) %>% 
  mutate(data = map(sample, function(x){
    log_path <- sprintf("%s/logs/210309_155222/%s/%s_arcas_genotype.log",isb_path,x,x)
    parse_arcas_log(log_path)
  })) %>%
  unnest(data) %>% 
  select(-data) %>% 
  separate(locus, into = c(NA, "locus"), sep = "-")
alignment_stats <- alignment_stats %>% 
  select(sample, all_reads_mapped) %>% 
  distinct()
```

```{r,fig.width = 16, fig.height = 6}
heatmap_df <- accuracy_df %>% 
  ungroup() %>% 
  left_join(cell_key, by = "sample") %>% 
  left_join(alignment_stats, by = "sample") %>% 
  mutate(n_cells = as.numeric(n_cells)) %>% 
  filter(grepl("-BL$", sample))
  
all_samples <- heatmap_df %>% pull(sample) %>% unique() %>% sort()
  
accuracy_heat <- heatmap_df %>% 
  filter(field == "field_2") %>%
  right_join(tibble(sample = all_samples), by = "sample") %>% 
  ggplot(aes(x=sample, y=genotyper, fill = accuracy))+
  geom_tile()+
  facet_grid(locus ~.) +
  scale_fill_viridis_c(na.value = "transparent") +
  theme(axis.text.x = element_text(angle = 90))

cells_heat <- heatmap_df %>% 
  select(sample, n_cells) %>% distinct() %>% 
  right_join(tibble(sample = all_samples), by = "sample") %>% 
  ggplot(aes(x=sample, y="a", fill = log10(n_cells)))+
  geom_tile()+
  scale_fill_viridis_c(na.value = "transparent") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  labs(y = "Cells")

reads_heat <- heatmap_df %>% 
  select(sample, all_reads_mapped) %>% distinct() %>% 
  right_join(tibble(sample = all_samples), by = "sample") %>% 
  ggplot(aes(x=sample, y="a", fill = log10(all_reads_mapped)))+
  geom_tile()+
  scale_fill_viridis_c(na.value = "transparent") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  labs(y = "Reads")

plot_grid(
  cells_heat+theme(axis.text.x = element_blank())+xlab(""),
  reads_heat+theme(axis.text.x = element_blank())+xlab(""),
  accuracy_heat+theme(axis.text.x = element_blank()),
  ncol = 1,
  rel_heights = c(2,2,16),
  align = "v", axis = "lr"
)
```

# Allele heatmaps

```{r}
accuracy_df %>% 
  unnest(reference) %>% 
  filter(locus == "A", field == "field_2") %>% 
  mutate(match = map2_dbl(reference, allele, function(x,y) sum(x %in% y, na.rm = t))) %>% 
  group_by(reference, genotyper) %>% 
  summarise(accuracy = mean(match)) %>% 
  ggplot(aes(x=reference, y=genotyper, fill = accuracy))+
  geom_tile()+
  # facet_grid(locus ~.) +
  scale_fill_viridis_c(na.value = "transparent") +
  theme(axis.text.x = element_text(angle = 90))

```


```{r, fig.width = 12, fig.height = 24}
accuracy_df %>% 
  unnest(reference) %>% 
  filter(field == "field_2") %>%
  filter(allele != "NA") %>% 
  # filter(field == "field_2") %>% 
  mutate(match = map2_dbl(reference, allele, function(x,y) sum(x %in% y, na.rm = T))) %>% 
  group_by(locus, reference, genotyper) %>% 
  summarise(accuracy = mean(match), n = n()) %>%
  ggplot(aes(x=reference, y = genotyper))+
  geom_point(aes(fill = accuracy, size = n), shape = 21, color = "black")+
  facet_wrap(~ locus, ncol = 1, scales = "free_x",  strip.position="right")+
  scale_fill_viridis_c(option = "C") +
  scale_size_continuous(breaks = c(5,25,50,100), range = c(1,10)) +
  theme_bw()+
  theme(axis.text.x = element_text(angle= 45, hjust = 1))


gg_allele_hla_accuracy(accuracy_df)
```

# Prediction frequency plots
```{r}
prediction_frequency <- all_hla %>% 
  select(-allele_id) %>% 
  pivot_longer(cols = field_1:field_3, names_to = "field", values_to = "allele") %>% 
  pivot_wider(names_from = "genotyper", values_from = "allele", values_fn = function(x) sum(!is.na(x)), values_fill = 0) %>% 
  # filter(invitro != 0) %>% 
  ### NEED TO DEAL WITH HLAMINER AMBIGOUS ALLELES
  # mutate_at(vars(arcasHLA:hlaminer), funs(./invitro)) %>% 
  mutate_at(vars(arcasHLA:hlaminer), ~./2) %>% 
  select(-invitro) %>% 
  ungroup() %>% 
  pivot_longer(cols = arcasHLA:hlaminer, names_to = "genotyper", values_to = "frequency")
prediction_frequency
```


```{r, fig.width = 6, fig.height = 8}
loci <- c("A","B","C","DPA1","DPB1","DQA1","DQB1","DRB1","DRB3","DRB4","DRB5")
prediction_frequency %>% 
  filter(frequency <= 1) %>% 
  filter(locus %in% loci) %>% 
  # complete(frequency = unique(frequency), fill = list(n=0)) %>% 
  mutate(count = frequency*2) %>% 
  ggplot(aes(x = genotyper))+
  geom_bar(aes(fill = factor(count)), position = "fill", color = "black", size = 0.25) +
  facet_grid(locus ~ field)+
  theme_bw()+
  scale_fill_viridis_d()+
  scale_y_continuous(breaks = c(0,0.5,1.0))+
  labs(x="",y="Proportion of samples",fill="Number of \npredicted alleles") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

```
























