---
title: "R Notebook"
output: html_notebook
---

```{r}
library(tidyverse)
library(ggcorrplot)
library(cowplot)
library(ggpubr)
```
# All sample heatmap
```{r, fig.width = 16, fig.height = 2}
for (l in c("A", "B", "C")){
  plt <- accuracy_df %>% 
    ungroup() %>% 
    filter(locus == l & grepl("BL$", sample) & field == "field_2") %>% 
    select(sample, genotyper, accuracy) %>% 
    pivot_wider(names_from = "genotyper", values_from = "accuracy", values_fill = 0) %>% 
    column_to_rownames("sample") %>% 
    select(rev(c("arcasHLA", "optitype", "phlat", "hlaminer"))) %>% 
    # t() %>% 
    ggcorrplot::ggcorrplot(
      
    ) +
    scale_fill_gradientn(colours = viridis::viridis(3))
  assign(sprintf("%s_plot", l), plt)
}

```
# Summarized heatmap

```{r}
plt <- accuracy_df %>% 
  group_by(locus, field, genotyper) %>% 
  summarise(mean_accuracy = mean(accuracy),
            inv_sd = 1-sd(accuracy)) %>% 
  ungroup() %>% 
  mutate(mean_accuracy = ifelse(mean_accuracy == 0, NA, mean_accuracy), 
         inv_sd = ifelse(mean_accuracy == 0, NA, inv_sd),
         locus = factor(locus, levels = sort(unique(locus), decreasing = T))) %>% 
  # select(-field) %>% 
  # pivot_wider(names_from = "genotyper", values_from = "accuracy") %>% 
  # column_to_rownames("locus") %>% 
  ggballoonplot(y = "locus", 
                x = "genotyper", 
                fill = "mean_accuracy", 
                size = "inv_sd", 
                facet.by = "field") +
  scale_fill_viridis_c(option = "C")
plt + 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90))+
  labs(x = "", fill = "Accuracy", size = "SD") +
  scale_size_continuous(breaks = c(0.9,0.8,0.7,0.6), 
                        labels = rev(c(0.9,0.8,0.7,0.6)),
                        range = c(0,10)
                        ) +
  ggtitle("4) Summarized heatmap")
```

# Original plot

```{r, fig.height = 6, fig.width = 4}
accuracy_df %>% 
  filter(grepl("^[ABC]", locus)) %>% 
  mutate(genotyper = fct_relevel(genotyper, "arcasHLA", "phlat", "optitype", "hlaminer")) %>% 
    ggplot(aes(x = genotyper, y = accuracy))+
    stat_summary(fun = mean, geom = "point", position = position_dodge(width=0.9)) +
    stat_summary(fun.data = mean_se, geom = "errorbar", position=position_dodge(width=0.9), width = 0.5) +
    theme_bw() +
    # facet_nested(.~locus+field, scales = "free_y") +
    facet_grid(field~locus, scales = "free") +
    coord_cartesian(ylim = c(0,1))+
    scale_y_continuous(n.breaks = 6)+
    labs(y = "Accuracy", x = "")+
    theme(axis.text.x = element_text(angle = 90))+
    theme(panel.spacing = unit(0.1,"line")) +
  ggtitle("1) Original Plot")
```

# Barplot

```{r, fig.width = 12, fig.height = 4}
accuracy_df %>% 
  filter(grepl("^[ABC]", locus)) %>% 
  mutate(genotyper = fct_relevel(genotyper, "arcasHLA", "phlat", "optitype", "hlaminer")) %>% 
  ggplot(aes(x = field, y = accuracy, fill = genotyper))+
  stat_summary(fun = mean, geom = "bar", position = "dodge", color = "black")  +
  stat_summary(fun.data = mean_se, geom = "errorbar", position=position_dodge(width = 0.9), width = 0.5) +
  theme_bw()  +
  scale_fill_brewer(palette = "Set1") +
  facet_grid(.~locus) +
  labs(x = "") +
  ggtitle("2) Adjacent bar plot")
```

# Individual heatmap

```{r}
cell_key <- tibble(line = system("wc -l ~/solomonb/covid/isb/extracted_barcodes/INCOV*", intern = T)) %>% 
  separate(line, into = c("n_cells", "file"), sep = " /") %>% 
  drop_na() %>% 
  mutate(sample = gsub("\\..*","",basename(file))) %>% 
  select(-file)
```

# Import arcas logs
```{r}
alignment_stats <- tibble(sample = isb_samples) %>% 
  mutate(data = map(sample, function(x){
    log_path <- sprintf("%s/logs/210309_155222/%s/%s_arcas_genotype.log",isb_path,x,x)
    parse_arcas_log(log_path)
  })) %>%
  unnest(data) %>% 
  select(-data) %>% 
  separate(locus, into = c(NA, "locus"), sep = "-")
alignment_stats <- alignment_stats %>% 
  select(sample, all_reads_mapped) %>% 
  distinct()
```

```{r,fig.width = 16, fig.height = 6}
heatmap_df <- accuracy_df %>% 
  ungroup() %>% 
  left_join(cell_key, by = "sample") %>% 
  left_join(alignment_stats, by = "sample") %>% 
  mutate(n_cells = as.numeric(n_cells)) %>% 
  filter(grepl("-BL$", sample))
  
all_samples <- heatmap_df %>% pull(sample) %>% unique() %>% sort()
  
accuracy_heat <- heatmap_df %>% 
  filter(field == "field_2") %>%
  right_join(tibble(sample = all_samples), by = "sample") %>% 
  ggplot(aes(x=sample, y=genotyper, fill = accuracy))+
  geom_tile()+
  facet_grid(locus ~.) +
  scale_fill_viridis_c(na.value = "transparent") +
  theme(axis.text.x = element_text(angle = 90))

cells_heat <- heatmap_df %>% 
  select(sample, n_cells) %>% distinct() %>% 
  right_join(tibble(sample = all_samples), by = "sample") %>% 
  ggplot(aes(x=sample, y="a", fill = log10(n_cells)))+
  geom_tile()+
  scale_fill_viridis_c(na.value = "transparent") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  labs(y = "Cells")

reads_heat <- heatmap_df %>% 
  select(sample, all_reads_mapped) %>% distinct() %>% 
  right_join(tibble(sample = all_samples), by = "sample") %>% 
  ggplot(aes(x=sample, y="a", fill = log10(all_reads_mapped)))+
  geom_tile()+
  scale_fill_viridis_c(na.value = "transparent") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  labs(y = "Reads")

plot_grid(
  cells_heat+theme(axis.text.x = element_blank())+xlab(""),
  reads_heat+theme(axis.text.x = element_blank())+xlab(""),
  accuracy_heat+theme(axis.text.x = element_blank()),
  ncol = 1,
  rel_heights = c(2,2,16),
  align = "v", axis = "lr"
)
```

# Allele heatmaps

```{r}
accuracy_df %>% 
  unnest(reference) %>% 
  filter(locus == "A", field == "field_2") %>% 
  mutate(match = map2_dbl(reference, allele, function(x,y) sum(x %in% y, na.rm = t))) %>% 
  group_by(reference, genotyper) %>% 
  summarise(accuracy = mean(match)) %>% 
  ggplot(aes(x=reference, y=genotyper, fill = accuracy))+
  geom_tile()+
  # facet_grid(locus ~.) +
  scale_fill_viridis_c(na.value = "transparent") +
  theme(axis.text.x = element_text(angle = 90))

```


```{r, fig.width = 12, fig.height = 6}
accuracy_df %>% 
  unnest(reference) %>% 
  filter(grepl("^[ABC]", locus), field == "field_2") %>%
  # filter(field == "field_2") %>% 
  mutate(match = map2_dbl(reference, allele, function(x,y) sum(x %in% y, na.rm = t))) %>% 
  group_by(locus, reference, genotyper) %>% 
  summarise(accuracy = mean(match), n = n()) %>%
  ggballoonplot(y = "genotyper", 
                x = "reference", 
                fill = "accuracy", 
                size = "n",
                facet.by = "locus") +
  scale_fill_viridis_c(option = "C") +
  scale_size_continuous(breaks = c(5,25,50,100), range = c(1,10)) +
  facet_wrap(~locus, scales = "free_x", ncol = 1)
```




























