---
title: "6) Accuracy on 3' 10X platform and bulk"
output: 
  html_notebook:
    toc: true
    toc_depth: 3
    toc_float: true
    code_folding: hide
---

# Reference info

- Publication: https://pubmed.ncbi.nlm.nih.gov/30518681/
- Ck and Sk represent a second time point for samples C and S, respectively
  - *"A second bone marrow aspiration was performed for 2 donors (Ck, Sk) (biological replicates) either 2 or 5 months after their first aspiration, respectively."*
- C1 and C2 represent the same sample split, technical replicates
  - From: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3396165
  - *"Description 	Technical replicate of C1"*


# Setup and import
```{r, message=F}
library(tidyverse)
library(tidymodels)
library(cowplot)
library(kknn)
library(here)

source(here("helper_functions/data_import_functions.R"))
source(here("helper_functions/figure_format_functions.R"))
source(here("helper_functions/calculation_functions.R"))
```

```{r, message = F}
invitro_hla <- invitro_import(path = here("data/pmid30518681/pmid30518681_scisco_hla.tsv"), 
                              exclude_comment_samples = T) %>% 
  mutate(sample = gsub("STU-HD-","",sample))

invitro_hla_c <- invitro_hla %>% filter(sample == "C")

# Expand results to technical and time replicates
invitro_hla <- bind_rows(
  invitro_hla,
  invitro_hla_c %>% mutate(sample = "Ck"),
  invitro_hla_c %>% mutate(sample = "C1"),
  invitro_hla_c %>% mutate(sample = "C2")
)

sample_metadata <- read_csv(here("data/pmid30518681/pmid30518681_samples.csv"))
pmid_3p_srr <- sample_metadata %>% filter(platform == "scRNA") %>% pull(Run)
pmid_3p_samples <- sample_metadata %>% filter(platform == "scRNA") %>% pull(sample)
pmid_bulk_srr <- sample_metadata %>% filter(platform == "bulkRNA") %>% pull(Run)
pmid_bulk_samples <- sample_metadata %>% filter(platform == "bulkRNA") %>% pull(sample)
```
### 3'
```{r}
pmid_3p_hla <- combine_HLA_import(
  path = here("data/pmid30518681/scRNA"),
  samples = pmid_3p_srr,
  invitro_path = NULL) %>%  # Will manually add invitro
  rename(Run = sample) %>% 
  left_join(sample_metadata, by = "Run") %>% 
  mutate(sample = ifelse(is.na(sample), original_name, sample)) %>% 
  select(sample, allele_id, allele, genotyper) %>% 
  bind_rows(invitro_hla) %>% 
  format_hla_table() %>% 
  filter(sample %in% pmid_3p_samples)
pmid_3p_hla
```

### Bulk
```{r}
pmid_bulk_hla <- combine_HLA_import(
  path = here("data/pmid30518681/bulkRNA"),
  samples = pmid_bulk_srr,
  invitro_path = NULL) %>%  # Will manually add invitro
  rename(Run = sample) %>% 
  left_join(sample_metadata, by = "Run") %>% 
  mutate(sample = ifelse(is.na(sample), original_name, sample)) %>% 
  select(sample, allele_id, allele, genotyper) %>% 
  bind_rows(invitro_hla) %>% 
  format_hla_table() %>% 
  filter(sample %in% pmid_bulk_samples)
pmid_bulk_hla
```



# Apply DRB345 KNN 

### Calculations
```{r}
arcas_log_3p_dir <- here("data/pmid30518681/scRNA/arcasHLA")
arcas_log_bulk_dir <- here("data/pmid30518681/bulkRNA/arcasHLA")

alignment_stats_3p_df <- hla_mapping_stats_import(pmid_3p_srr, arcas_log_3p_dir)
alignment_stats_bulk_df <- hla_mapping_stats_import(pmid_bulk_srr, arcas_log_bulk_dir)
```

```{r}
# Get DRB345/DRB1 ratios
drb345_ratios_3p <- get_drb345_ratios(alignment_stats_3p_df) %>% 
  rename("Run"="sample") %>% 
  left_join(sample_metadata %>% select(Run, sample), by = "Run") %>% 
  select(-Run)
drb345_ratios_bulk <- get_drb345_ratios(alignment_stats_bulk_df) %>% 
  rename("Run"="sample") %>% 
  left_join(sample_metadata %>% select(Run, sample), by = "Run") %>% 
  select(-Run)
```

```{r, message = F}
pmid_copy_number <- get_drb345_copy_number(here("data/pmid30518681/pmid30518681_scisco_hla.tsv")) %>% 
  mutate(sample = gsub(".*HD-","",sample))

# Duplicate in vitro typing for replicate samples Ck, C1, C2
pmid_copy_number_c <- pmid_copy_number %>% filter(sample == "C")

pmid_copy_number <- bind_rows(
  pmid_copy_number,
  pmid_copy_number_c %>% mutate(sample = "Ck"),
  pmid_copy_number_c %>% mutate(sample = "C1"),
  pmid_copy_number_c %>% mutate(sample = "C2")
)
```

```{r}
# Combine DRB345 sequencing ratios and ground truth copy numbers
drb345_df_3p <- create_drb345_df(drb345_ratios_3p, pmid_copy_number)
drb345_df_bulk <- create_drb345_df(drb345_ratios_bulk, pmid_copy_number)
```

```{r}
# Apply KNN model
drb_model <- readRDS(here("3_DRB/drb345_knn_model.RDS"))
drb345_3p_predicted <- apply_drb345_knn(drb345_df_3p, drb_model)
drb345_bulk_predicted <- apply_drb345_knn(drb345_df_bulk, drb_model)
```

### ROC plots and AUC calculations
```{r}
auc_estimator <- "hand_till"
# Generate ROC plots and AUC
drb345_3p_auc <- drb345_3p_predicted %>% 
  roc_auc(truth = copy_number, .pred_0:.pred_2, estimator = auc_estimator) %>% 
  pull(.estimate) %>% 
  round(.,3)
drb345_ROC_3p <- gg_multilevel_roc(drb345_3p_predicted) +
  geom_label(aes(label = sprintf("AUC: %s", drb345_3p_auc), x=0.8, y=0.05, color = NULL), color = "black", size = 3)
drb345_ROC_3p

# None of the bulk RNA samples have 2 alleles for a single DRB345. 
# ROC will fail to generate without all possible allele counts {0,1,2} in ground truth
# Will add in place holder {2}, which will allow ROC to proceed
# Will exclude class-{2} ROC plot
# class-{0} and class-{1} plots are not affected by the "pseudotruth" addition
drb345_bulk_predicted_pseudo <- drb345_bulk_predicted %>% 
  mutate_if(is.factor, as.character) %>% 
  rbind(., list("2_copy_pseudotruth",0,0,0,"DRB3","2","2",0,0,1)) %>% 
  mutate_if(is.character, factor)

drb345_bulk_auc <- drb345_bulk_predicted_pseudo %>% 
  filter(copy_number != "2") %>% # Removing {2} pseudotruth here
  roc_auc(truth = copy_number, .pred_0:.pred_2, estimator = auc_estimator) %>% 
  pull(.estimate) %>% 
  round(.,3)

# Proceed normally from here
drb345_ROC_bulk <-  drb345_bulk_predicted_pseudo %>% 
  roc_curve(truth = copy_number, .pred_0:.pred_2) %>% 
  autoplot() %>% 
  (function(x) ggplot_build(x)$data[[1]]) %>% 
  mutate(n_alleles = fct_recode(PANEL, "0"="1", "1"="2", "2"="3")) %>% 
  filter(n_alleles != "2") %>% # Removing {2} pseudotruth here
  ggplot(aes(x=x,y=y,color=n_alleles))+
  geom_path()+
  geom_abline(slope=1,intercept=0, linetype="dotted")+
  theme_bw() +
  scale_color_brewer(palette = "Set1") +
  labs(x = "1 - Specificity", y = "Sensitivity", color = "Number\nof alleles") +
  geom_label(aes(label = sprintf("AUC: %s", drb345_bulk_auc), x=0.8, y=0.05, color = NULL), color = "black", size = 3)

drb345_ROC_bulk

```
```{r}
drb345_bulk_predicted
```



```{r}
# Demonstrate that pseudotruth row doesn't affect other ROC classes
# Reformats outcome as a binary classifier. 
# Event_level specifies {1} outcome as the base event, so plot should match {1} plot above
drb345_bulk_predicted %>% 
  mutate(copy_number = factor(copy_number, levels = c("0","1"))) %>% 
  roc_curve(truth = copy_number, .pred_1, event_level = "second") %>% 
  autoplot()
```

### DRB345 ratio distribution plots

```{r, message = F, warning = F}
drb345_df_5p <- read_csv(here("3_DRB/isb_drb345_df.csv"))
drb345_ratio_plt_3p <- gg_drb345_ratio(drb345_df_3p, drb345_df_5p)
drb345_ratio_plt_bulk <- gg_drb345_ratio(drb345_df_bulk, drb345_df_5p)
drb345_ratio_plt_3p
drb345_ratio_plt_bulk
```

```{r, fig.width=8, fig.height=6}
plot_grid(
  drb345_ROC_3p,
  drb345_ratio_plt_3p,
  drb345_ROC_bulk,
  drb345_ratio_plt_bulk,
  nrow = 2,
  ncol = 2,
  labels = LETTERS[1:4],
  align = "h",
  axis = "bt"
)
```

### Predicted tally vs. ground truth
```{r, fig.width = 3, fig.height=5}
post_tally_3p_plt <- drb345_3p_predicted %>% 
  rename("copy_invitro" = copy_number, "copy_predicted" = .pred_class) %>% 
  pivot_longer(contains("copy_"), names_to = "genotyper", values_to = "count", values_transform = list(count = function(x) as.numeric(as.character(x))), names_prefix = "copy_") %>% 
  mutate(field = " ") %>% 
  gg_hla_prediction_tally()
post_tally_3p_plt

post_tally_bulk_plt <- drb345_bulk_predicted %>% 
  rename("copy_invitro" = copy_number, "copy_predicted" = .pred_class) %>% 
  pivot_longer(contains("copy_"), names_to = "genotyper", values_to = "count", values_transform = list(count = function(x) as.numeric(as.character(x))), names_prefix = "copy_") %>% 
  mutate(field = " ") %>% 
  gg_hla_prediction_tally()
post_tally_bulk_plt
```

#### Update DRB345 alleles based on copy number modeling

```{r}
all_hla_drb345_3p_filtered <- filter_drb_in_all_hla(drb_predictions = drb345_3p_predicted, 
                                                    all_hla_df = pmid_3p_hla, 
                                                    samples = pmid_3p_samples)
all_hla_drb345_bulk_filtered <- filter_drb_in_all_hla(drb_predictions = drb345_bulk_predicted, 
                                                      all_hla_df = pmid_bulk_hla, 
                                                      samples = pmid_bulk_samples)

# saveRDS(all_hla_drb345_3p_filtered, here("8_platform_compare/all_hla_drb345_filtered_pmid_scRNA.RDS"))
# saveRDS(all_hla_drb345_bulk_filtered, here("8_platform_compare/all_hla_drb345_filtered_pmid_bulkRNA.RDS"))

# all_hla_drb345_3p_filtered <- readRDS(here("8_platform_compare/all_hla_drb345_filtered_pmid_scRNA.RDS"))
# all_hla_drb345_bulk_filtered <- readRDS(here("8_platform_compare/all_hla_drb345_filtered_pmid_bulkRNA.RDS"))
```

# Plots and tables

```{r, fig.width = 5, fig.height = 8}
prediction_tally_3p_plt <- all_hla_drb345_3p_filtered %>% 
  allele_tally() %>% 
  filter(genotyper != "invitro") %>% 
  gg_hla_prediction_tally()
prediction_tally_3p_plt

prediction_tally_bulk_plt <- all_hla_drb345_bulk_filtered %>% 
  allele_tally() %>% 
  filter(genotyper != "invitro") %>% 
  gg_hla_prediction_tally()
prediction_tally_bulk_plt
```

### Success
```{r, fig.width = 5, fig.height = 4}
success_3p_df <- calculate_all_hla_accuracy(all_hla_drb345_3p_filtered, method = "success")
success_3p_plt <- calculate_summary_df(success_3p_df) %>%
  gg_summary_hla_accuracy(color_label = "Success")
success_3p_plt

success_bulk_df <- calculate_all_hla_accuracy(all_hla_drb345_bulk_filtered, method = "success")
success_bulk_plt <- calculate_summary_df(success_bulk_df) %>%
  gg_summary_hla_accuracy(color_label = "Success")
success_bulk_plt
```


### Accuracy
```{r, fig.width = 5, fig.height = 4}
accuracy_3p_df <- calculate_all_hla_accuracy(all_hla_drb345_3p_filtered, method = "accuracy")
accuracy_3p_plt <- calculate_summary_df(accuracy_3p_df) %>%
  gg_summary_hla_accuracy(color_label = "Accuracy")
accuracy_3p_plt

accuracy_bulk_df <- calculate_all_hla_accuracy(all_hla_drb345_bulk_filtered, method = "accuracy")
accuracy_bulk_plt <- calculate_summary_df(accuracy_bulk_df) %>%
  gg_summary_hla_accuracy(color_label = "Accuracy")
accuracy_bulk_plt
```

### Combined plots
```{r,fig.height = 6, fig.width = 16}
plt_3p_heatmaps <- plot_grid(
  success_3p_plt,
  accuracy_3p_plt,
  ncol = 1,
  labels = c("B","C")
)
plt_3p <- plot_grid(
  prediction_tally_3p_plt,
  plt_3p_heatmaps,
  labels = c("A",NA)
)

plt_bulk_heatmaps <- plot_grid(
  success_bulk_plt,
  accuracy_bulk_plt,
  ncol = 1,
  labels = c("E","F")
)
plt_bulk <- plot_grid(
  prediction_tally_bulk_plt,
  plt_bulk_heatmaps,
  labels = c("D",NA)
)

plot_grid(
  plt_3p,
  plt_bulk,
  nrow=1
)
```
```{r,fig.height = 3, fig.width = 8}
plt_fig_supp <- plot_grid(accuracy_3p_plt,
          accuracy_bulk_plt,
          nrow = 1,
          labels = LETTERS[1:2])
plt_fig_supp
```
```{r}
save_plot(here("figures_pdf/v2/s1_platforms.pdf"), plt_fig_supp, base_height = 3, base_width = 8)
```

### Accuracy tables

```{r}
p3_accuracy_flextable <- calculate_summary_df(accuracy_3p_df) %>% 
  flex_summary_hla_accuracy() %>% 
  fontsize(size = 8, part = "all")
# saveRDS(p3_accuracy_flextable, here("tables/p3_accuracy_flextable.RDS"))
print(p3_accuracy_flextable)
```

```{r}
bulk_accuracy_flextable <- calculate_summary_df(accuracy_bulk_df) %>% 
  flex_summary_hla_accuracy() %>% 
  fontsize(size = 8, part = "all") 
# saveRDS(bulk_accuracy_flextable, here("tables/bulk_accuracy_flextable.RDS"))
print(bulk_accuracy_flextable)
```

# Save all figure objects 
```{r}
listN <- function(...){
    anonList <- list(...)
    names(anonList) <- as.character(substitute(list(...)))[-1]
    anonList
}

saveRDS(
  listN(
    drb345_ROC_3p,
    drb345_ratio_plt_3p,
    drb345_ROC_bulk,
    drb345_ratio_plt_bulk,
    success_3p_plt,
    accuracy_3p_plt,
    prediction_tally_3p_plt,
    success_bulk_plt,
    accuracy_bulk_plt,
    prediction_tally_bulk_plt
  ),
  here("figures_object/6_platforms.RDS")
)

```

# OLD WORK

<!-- ### Prediction frequency -->
<!-- ```{r, fig.width = 5, fig.height = 8} -->
<!-- prediction_3p_frequency <-   allele_tally(pmid_3p_hla)   -->
<!-- predictions_3p_plt <- gg_hla_prediction_tally(prediction_3p_frequency) -->
<!-- predictions_3p_plt -->
<!-- ``` -->

<!-- ```{r, fig.width = 5, fig.height = 8} -->
<!-- prediction_bulk_frequency <-   allele_tally(pmid_bulk_hla)   -->
<!-- predictions_bulk_plt <- gg_hla_prediction_tally(prediction_bulk_frequency) -->
<!-- predictions_bulk_plt -->
<!-- ``` -->

<!-- ### Success -->
<!-- ```{r, fig.width = 5, fig.height = 4} -->
<!-- success_3p_df <- calculate_all_hla_accuracy(pmid_3p_hla, method = "success") -->
<!-- success_3p_plt <- calculate_summary_df(success_3p_df) %>%  -->
<!--   gg_summary_hla_accuracy(color_label = "Success") -->
<!-- success_3p_plt -->
<!-- ``` -->

<!-- ```{r, fig.width = 5, fig.height = 4} -->
<!-- success_bulk_df <- calculate_all_hla_accuracy(pmid_bulk_hla, method = "success") -->
<!-- success_bulk_plt <- calculate_summary_df(success_bulk_df) %>%  -->
<!--   gg_summary_hla_accuracy(color_label = "Success") -->
<!-- success_bulk_plt -->
<!-- ``` -->

<!-- ### Accuracy -->
<!-- ```{r, fig.width = 5, fig.height = 4} -->
<!-- accuracy_3p_df <- calculate_all_hla_accuracy(pmid_3p_hla, method = "accuracy") -->
<!-- accuracy_3p_plt <- calculate_summary_df(accuracy_3p_df) %>%  -->
<!--   gg_summary_hla_accuracy(color_label = "Accuracy") -->
<!-- accuracy_3p_plt -->
<!-- ``` -->

<!-- ```{r, fig.width = 5, fig.height = 4} -->
<!-- accuracy_bulk_df <- calculate_all_hla_accuracy(pmid_bulk_hla, method = "accuracy") -->
<!-- accuracy_bulk_plt <- calculate_summary_df(accuracy_bulk_df) %>%  -->
<!--   gg_summary_hla_accuracy(color_label = "Accuracy") -->
<!-- accuracy_bulk_plt -->
<!-- ``` -->

<!-- ### Combined plot -->
<!-- ```{r,fig.height = 6, fig.width = 16} -->
<!-- combine_plt <- cowplot::plot_grid( -->
<!--   cowplot::plot_grid( -->
<!--     predictions_3p_plt, -->
<!--     cowplot::plot_grid( -->
<!--       success_3p_plt, -->
<!--       accuracy_3p_plt, -->
<!--       ncol=1 -->
<!--     ), -->
<!--     nrow=1 -->
<!--   ), -->
<!--   cowplot::plot_grid( -->
<!--     predictions_bulk_plt, -->
<!--     cowplot::plot_grid( -->
<!--       success_bulk_plt, -->
<!--       accuracy_bulk_plt, -->
<!--       ncol=1 -->
<!--     ), -->
<!--     nrow=1 -->
<!--   ), -->
<!--   nrow=1 -->
<!-- ) -->

<!-- cowplot::plot_grid( -->
<!--   cowplot::plot_grid( -->
<!--     NULL, -->
<!--     NULL, -->
<!--     nrow=1, -->
<!--     labels = c("3' scRNA sequencing", "Bulk RNA sequencing") -->
<!--   ), -->
<!--   combine_plt, -->
<!--   ncol=1, -->
<!--   rel_heights = c(0.05,0.95) -->
<!-- ) -->
<!-- ``` -->


<!-- ```{r, fig.width = 8, fig.height = 4} -->
<!-- c1_a <- accuracy_df%>%  -->
<!--   filter(grepl("^[ABC]", locus)) %>%  -->
<!--   gg_accuracy() -->

<!-- dp_a <- accuracy_df %>%  -->
<!--   filter(grepl("^DP", locus) & genotyper %in% c("arcasHLA", "hlaminer"),) %>%  -->
<!--   gg_accuracy() -->

<!-- dq_a <- accuracy_df %>%  -->
<!--   filter(grepl("^DQ", locus) & genotyper %in% c("arcasHLA", "hlaminer", "phlat"),) %>%  -->
<!--   gg_accuracy() -->

<!-- drb1_a <- accuracy_df %>%  -->
<!--   filter(grepl("^DRB1", locus) & genotyper %in% c("arcasHLA", "hlaminer", "phlat"),) %>%  -->
<!--   gg_accuracy() -->

<!-- drb345_a <- accuracy_df %>%  -->
<!--   filter(grepl("^DRB[345]", locus) & genotyper %in% c("arcasHLA", "hlaminer"),) %>%  -->
<!--   gg_accuracy() -->

<!-- c2_a <- dp_a| -->
<!--   strip_y(dq_a)| -->
<!--   strip_y(drb1_a)| -->
<!--   strip_y(drb345_a)| -->
<!--   plot_layout(widths = c(6,6,3,9)) -->

<!-- c1_a | c2_a | plot_layout(widths = c(4,13)) -->
<!-- ``` -->



